// @ccms/core/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var containerStartTime = Date.now();
console.i18n("ms.core.ioc.initialize", { scope: global.scope });
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
console.i18n("ms.core.ioc.completed", { scope: global.scope, time: (Date.now() - containerStartTime) / 1000 });
var yaml = require("js-yaml");
var http_1 = require("@ccms/common/dist/http");
var fs = require("@ccms/common/dist/fs");
var version_1 = require("@ccms/common/dist/version");
var UUID = Java.type('java.util.UUID');
var MiaoScriptAPI = Java.type('pw.yumc.MiaoScript.api.MiaoScriptAPI');
var MiaoScriptCore = /** @class */ (function () {
    function MiaoScriptCore() {
    }
    MiaoScriptCore_1 = MiaoScriptCore;
    MiaoScriptCore.prototype.enable = function () {
        var _this = this;
        process.emit('core.before.enable');
        this.loadServerConsole();
        try {
            MiaoScriptAPI.setPluginManager(this.pluginManager);
        }
        catch (error) {
        }
        this.loadPlugins();
        process.emit('core.after.enable');
        console.i18n("ms.core.engine.completed", {
            loader: base.version,
            version: 'v' + global.ScriptEngineVersion,
            time: (Date.now() - global.ScriptEngineStartTime) / 1000
        });
        return function () { return _this.disable(); };
    };
    MiaoScriptCore.prototype.loadServerConsole = function () {
        process.emit('core.before.load.console');
        //@ts-ignore
        global.setGlobal('console', new this.Console(), { writable: false, configurable: false });
        process.emit('core.after.load.console');
    };
    MiaoScriptCore.prototype.loadPlugins = function () {
        process.emit('core.before.load.plugins');
        var loadPluginStartTime = new Date().getTime();
        console.i18n("ms.core.plugin.initialize");
        this.pluginManager.scan(this.pluginFolder);
        this.pluginManager.build();
        this.pluginManager.load(this.pluginManager.getPlugins());
        this.pluginManager.enable(this.pluginManager.getPlugins());
        console.i18n("ms.core.plugin.completed", { time: (new Date().getTime() - loadPluginStartTime) / 1000 });
        process.emit('core.after.load.plugins');
    };
    MiaoScriptCore.prototype.disable = function () {
        process.emit('core.before.disable');
        var disableStartTime = Date.now();
        console.i18n("ms.core.engine.disable");
        this.pluginManager.disable(this.pluginManager.getPlugins());
        this.taskManager.disable();
        process.emit('core.after.disable');
        loadCoreScript('exit');
        process.emit('core.before.exit');
        process.exit(0);
        console.i18n("ms.core.engine.disable.finish", {
            loader: base.version,
            version: 'v' + global.ScriptEngineVersion,
            time: (new Date().getTime() - disableStartTime) / 1000
        });
        process.emit('core.after.exit');
    };
    var MiaoScriptCore_1;
    tslib_1.__decorate([
        (0, container_1.Autowired)(api_1.server.Console),
        tslib_1.__metadata("design:type", Object)
    ], MiaoScriptCore.prototype, "Console", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(api_1.plugin.PluginFolder),
        tslib_1.__metadata("design:type", String)
    ], MiaoScriptCore.prototype, "pluginFolder", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", api_1.task.TaskManager)
    ], MiaoScriptCore.prototype, "taskManager", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", api_1.plugin.PluginManager)
    ], MiaoScriptCore.prototype, "pluginManager", void 0);
    MiaoScriptCore = MiaoScriptCore_1 = tslib_1.__decorate([
        (0, container_1.provideSingleton)(MiaoScriptCore_1)
    ], MiaoScriptCore);
    return MiaoScriptCore;
}());
function detectServer() {
    try {
        Java.type("org.bukkit.Bukkit");
        return api_1.constants.ServerType.Bukkit;
    }
    catch (ex) {
    }
    try {
        Java.type("org.spongepowered.api.Sponge");
        return api_1.constants.ServerType.Sponge;
    }
    catch (ex) {
    }
    try {
        Java.type("cn.nukkit.Nukkit");
        return api_1.constants.ServerType.Nukkit;
    }
    catch (ex) {
    }
    try {
        Java.type("net.md_5.bungee.api.ProxyServer");
        return api_1.constants.ServerType.Bungee;
    }
    catch (ex) {
    }
    try {
        Java.type("org.springframework.boot.SpringApplication");
        return api_1.constants.ServerType.Spring;
    }
    catch (ex) {
    }
    throw Error('Unknow Server Type...');
}
function loadCoreScript(name) {
    try {
        var scriptname = name + (global.debug ? '-debug' : '');
        engineLoad({
            script: http_1.default.get("https://mscript.yumc.pw/api/plugin/download/name/".concat(scriptname)),
            name: "core/".concat(scriptname, ".js")
        });
    }
    catch (error) {
        if (global.debug) {
            console.debug(error);
            console.ex(error);
        }
    }
}
function loadMiaoScriptConfig() {
    var configFile = fs.concat(root, 'config.yml');
    if (!fs.exists(configFile)) {
        global.ScriptEngineConfig = base.save(configFile, yaml.dump({
            uuid: UUID.randomUUID().toString(),
            slow_execute: 50
        }));
    }
    else {
        global.ScriptEngineConfig = yaml.load(base.read(configFile));
    }
    global.ScriptSlowExecuteTime = global.ScriptEngineConfig.slow_execute || 50;
}
function createCore() {
    var corePackageStartTime = new Date().getTime();
    container_1.DefaultContainer.bind(container_1.ContainerInstance).toConstantValue(container_1.DefaultContainer);
    container_1.DefaultContainer.bind(api_1.plugin.PluginInstance).toConstantValue(base.getInstance());
    container_1.DefaultContainer.bind(api_1.plugin.PluginFolder).toConstantValue('plugins');
    var type = detectServer();
    process.emit('core.before.initialize.detect');
    console.i18n("ms.core.initialize.detect", { scope: global.scope, type: type });
    container_1.DefaultContainer.bind(api_1.server.ServerType).toConstantValue(type);
    container_1.DefaultContainer.bind(api_1.server.ServerChecker).toSelf().inSingletonScope();
    container_1.DefaultContainer.bind(api_1.server.NativePluginManager).toSelf().inSingletonScope();
    process.emit('core.after.initialize.detect');
    process.emit('core.before.package.initialize');
    console.i18n("ms.core.package.initialize", { scope: global.scope, type: type });
    require("".concat(global.scope, "/").concat(type)).default(container_1.DefaultContainer);
    require("".concat(global.scope, "/plugin"));
    container_1.DefaultContainer.load((0, container_1.buildProviderModule)());
    console.i18n("ms.core.package.completed", { scope: global.scope, type: type, time: (Date.now() - corePackageStartTime) / 1000 });
    process.emit('core.after.package.initialize');
    return container_1.DefaultContainer.get(MiaoScriptCore);
}
function initialize() {
    process.emit('core.before.initialize');
    loadMiaoScriptConfig();
    global.ScriptEngineVersion = require('../package.json').version;
    global.setGlobal('loadCoreScript', loadCoreScript);
    loadCoreScript('initialize');
    try {
        var core = createCore();
        return version_1.VersionUtils.isGreaterOrEqual(base.version, '0.22.0') ? core : core.enable();
    }
    catch (error) {
        var core = { enable: function () { return function () { return console.i18n('ms.core.engine.disable.abnormal'); }; } };
        console.i18n("core.initialize.error", { error: error });
        api_1.console.getStackTrace(error, false).forEach(function (line) { return console.log(line); });
        process.emit('core.initialize.error');
        return version_1.VersionUtils.isGreaterOrEqual(base.version, '0.22.0') ? core : core.enable();
    }
    finally {
        process.emit('core.after.initialize');
    }
}
exports.default = initialize();
