// @ccms/bukkit/dist/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BukkitCommand = void 0;
var tslib_1 = require("tslib");
require("@ccms/nashorn");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var reflect = require("@ccms/common/dist/reflect");
var Bukkit = Java.type('org.bukkit.Bukkit');
var TabCompleter = Java.type('org.bukkit.command.TabCompleter');
var PluginCommand = Java.type('org.bukkit.command.PluginCommand');
var CommandExecutor = Java.type('org.bukkit.command.CommandExecutor');
var BukkitCommand = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitCommand, _super);
    function BukkitCommand() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BukkitCommand.prototype.init = function () {
        this.commandMap = reflect.on(Bukkit.getPluginManager()).get('commandMap').get();
        this.knownCommands = reflect.on(this.commandMap).get('knownCommands').get();
    };
    BukkitCommand.prototype.create = function (plugin, command) {
        var cmd = this.commandMap.getCommand(command);
        if (cmd instanceof PluginCommand) {
            return cmd;
        }
        ;
        cmd = reflect.on(PluginCommand).create(command, this.pluginInstance).get();
        this.commandMap.register(plugin.description.name, cmd);
        return cmd;
    };
    BukkitCommand.prototype.remove = function (plugin, command) {
        var cmd = this.commandMap.getCommand(command);
        if (cmd instanceof PluginCommand) {
            cmd.unregister(this.commandMap);
            this.knownCommands.remove(command);
        }
    };
    BukkitCommand.prototype.tabComplete = function (sender, input, index) {
        return Java.from(this.commandMap.tabComplete(sender, input));
    };
    BukkitCommand.prototype.onCommand = function (plugin, command, executor) {
        // 必须指定需要实现的接口类型 否则MOD服会报错
        command.setExecutor(new CommandExecutor({
            onCommand: _super.prototype.setExecutor.call(this, plugin, command, executor)
        }));
    };
    BukkitCommand.prototype.onTabComplete = function (plugin, command, tabCompleter) {
        // 必须指定需要实现的接口类型 否则MOD服会报错
        command.setTabCompleter(new TabCompleter({
            onTabComplete: _super.prototype.setTabCompleter.call(this, plugin, command, tabCompleter)
        }));
    };
    tslib_1.__decorate([
        (0, container_1.inject)(api_1.plugin.PluginInstance),
        tslib_1.__metadata("design:type", Object)
    ], BukkitCommand.prototype, "pluginInstance", void 0);
    tslib_1.__decorate([
        (0, container_1.postConstruct)(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], BukkitCommand.prototype, "init", null);
    BukkitCommand = tslib_1.__decorate([
        (0, container_1.provideSingleton)(api_1.command.Command)
    ], BukkitCommand);
    return BukkitCommand;
}(api_1.command.Command));
exports.BukkitCommand = BukkitCommand;
