// @ccms/bukkit/dist/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BukkitEvent = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var reflect = require("@ccms/common/dist/reflect");
var URL = Java.type('java.net.URL');
var Bukkit = Java.type("org.bukkit.Bukkit");
var Event = Java.type("org.bukkit.event.Event");
var Modifier = Java.type("java.lang.reflect.Modifier");
var Listener = Java.type("org.bukkit.event.Listener");
var EventPriority = Java.type("org.bukkit.event.EventPriority");
var EventExecutor = Java.type("org.bukkit.plugin.EventExecutor");
var BukkitEvent = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitEvent, _super);
    function BukkitEvent() {
        return _super.call(this, 'org/bukkit/event') || this;
    }
    BukkitEvent.prototype.getJarFile = function (resource) {
        try {
            return _super.prototype.getJarFile.call(this, 'org/bukkit/Bukkit.class', Bukkit.class.classLoader);
        }
        catch (error) {
            // 兼容 LoliServer 的 Bukkit 包无法获取的问题
            var ModList = Java.type('net.minecraftforge.fml.ModList').get();
            var forgeFile = ModList.getModFileById("forge").getFile().getFilePath();
            var jarPath = "jar:file:".concat(forgeFile, "!/org/bukkit/Bukkit.class");
            return new URL(jarPath).openConnection().jarFile;
        }
    };
    BukkitEvent.prototype.isValidEvent = function (clazz) {
        // 继承于 org.bukkit.event.Event
        return Event.class.isAssignableFrom(clazz) &&
            // 访问符为Public
            Modifier.isPublic(clazz.getModifiers()) &&
            // 不是抽象类
            !Modifier.isAbstract(clazz.getModifiers());
    };
    BukkitEvent.prototype.register = function (eventCls, exec, priority, ignoreCancel) {
        var listener = new Listener({});
        Bukkit.pluginManager.registerEvent(eventCls, listener, EventPriority[priority], new EventExecutor({
            execute: exec
        }), this.pluginInstance, ignoreCancel);
        return listener;
    };
    BukkitEvent.prototype.unregister = function (event, listener) {
        reflect.on(event).call('getHandlerList').get().unregister(listener);
    };
    tslib_1.__decorate([
        (0, container_1.inject)(api_1.plugin.PluginInstance),
        tslib_1.__metadata("design:type", Object)
    ], BukkitEvent.prototype, "pluginInstance", void 0);
    BukkitEvent = tslib_1.__decorate([
        (0, container_1.provideSingleton)(api_1.event.Event),
        tslib_1.__metadata("design:paramtypes", [])
    ], BukkitEvent);
    return BukkitEvent;
}(api_1.event.Event));
exports.BukkitEvent = BukkitEvent;
