// @ccms/bukkit/dist/item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BukkitItemBuilder = exports.BukkitItem = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var item_1 = require("./internal/item");
var Material = Java.type('org.bukkit.Material');
var ItemStack = Java.type('org.bukkit.inventory.ItemStack');
var BukkitItem = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitItem, _super);
    function BukkitItem() {
        var _this = _super.call(this) || this;
        _this.reflect();
        return _this;
    }
    BukkitItem.prototype.builder = function () {
        return new BukkitItemBuilder();
    };
    BukkitItem.prototype.toJson = function (item) {
        var nbt = new this.NBTTagCompound();
        return this.CraftItemStack.asNMSCopy(item)[this.nmsSaveNBTMethodName](nbt).toString();
    };
    BukkitItem.prototype.fromJson = function (json) {
        return this.CraftItemStack.asBukkitCopy(new this.nmsItemStack(this.MojangsonParser[this.mpParseMethodName](json)));
    };
    BukkitItem.prototype.obcCls = function (name) {
        return base.getClass(['org.bukkit.craftbukkit', this.nmsVersion, name].join('.'));
    };
    BukkitItem.prototype.nmsCls = function (name) {
        return base.getClass(['net.minecraft.server', this.nmsVersion, name].join('.'));
    };
    BukkitItem.prototype.reflect = function () {
        var e_1, _a, e_2, _b, e_3, _c;
        try {
            var Bukkit = Java.type('org.bukkit.Bukkit');
            // @ts-ignore
            this.nmsVersion = Bukkit.getServer().class.name.split('.')[3];
            var CraftItemStackClass = this.obcCls('inventory.CraftItemStack');
            this.CraftItemStack = Java.type(CraftItemStackClass.getName());
            // @ts-ignore
            var asNMSCopyMethod = CraftItemStackClass.getMethod('asNMSCopy', ItemStack.class);
            var nmsItemStackClass = asNMSCopyMethod.getReturnType();
            this.nmsItemStack = Java.type(nmsItemStackClass.getName());
            var nmsNBTTagCompoundClass = undefined;
            try {
                for (var _d = tslib_1.__values(Java.from(nmsItemStackClass.getMethods())), _e = _d.next(); !_e.done; _e = _d.next()) {
                    var method = _e.value;
                    var rt = method.getReturnType();
                    if (method.getParameterTypes().length == 0 && "NBTTagCompound" == rt.getSimpleName()) {
                        nmsNBTTagCompoundClass = rt;
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                }
                finally { if (e_1) throw e_1.error; }
            }
            this.NBTTagCompound = Java.type(nmsNBTTagCompoundClass.getName());
            try {
                for (var _f = tslib_1.__values(Java.from(nmsItemStackClass.getMethods())), _g = _f.next(); !_g.done; _g = _f.next()) {
                    var method = _g.value;
                    var params = method.getParameterTypes();
                    var rt = method.getReturnType();
                    if (params.length == 1 && "NBTTagCompound" == params[0].getSimpleName() && "NBTTagCompound" == rt.getSimpleName()) {
                        this.nmsSaveNBTMethodName = method.getName();
                        break;
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_g && !_g.done && (_b = _f.return)) _b.call(_f);
                }
                finally { if (e_2) throw e_2.error; }
            }
            try {
                this.MojangsonParser = this.nmsCls('MojangsonParser').static;
            }
            catch (error) {
                this.MojangsonParser = Java.type('net.minecraft.nbt.MojangsonParser');
            }
            try {
                for (var _h = tslib_1.__values(Java.from(this.MojangsonParser.class.getMethods())), _j = _h.next(); !_j.done; _j = _h.next()) {
                    var method = _j.value;
                    var params = method.getParameterTypes();
                    var rt = method.getReturnType();
                    if (params.length == 1 && "String" == params[0].getSimpleName() && "NBTTagCompound" == rt.getSimpleName()) {
                        this.mpParseMethodName = method.getName();
                    }
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_j && !_j.done && (_c = _h.return)) _c.call(_h);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        catch (error) {
            console.log('Bukkit 物品管理器初始化失败:', error);
            if (global.debug) {
                console.ex(error);
            }
        }
    };
    BukkitItem = tslib_1.__decorate([
        (0, container_1.provideSingleton)(api_1.item.Item),
        tslib_1.__metadata("design:paramtypes", [])
    ], BukkitItem);
    return BukkitItem;
}(api_1.item.Item));
exports.BukkitItem = BukkitItem;
var BukkitItemBuilder = /** @class */ (function () {
    function BukkitItemBuilder() {
    }
    BukkitItemBuilder.prototype.from = function (itemStack) {
        this.itemStack = itemStack;
        this.itemMeta = this.itemStack.getItemMeta();
        return this;
    };
    BukkitItemBuilder.prototype.create = function (type) {
        var material = undefined;
        if (typeof type == 'number') {
            type = item_1.itemIds[type];
        }
        material = Material[type] || Material[Material['LEGACY_PREFIX'] + type];
        this.itemStack = new ItemStack(material);
        this.itemMeta = this.itemStack.getItemMeta();
        return this;
    };
    BukkitItemBuilder.prototype.name = function (name) {
        this.itemMeta.setDisplayName(name);
        return this;
    };
    BukkitItemBuilder.prototype.lore = function () {
        var lores = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            lores[_i] = arguments[_i];
        }
        this.itemMeta.setLore(lores);
        return this;
    };
    BukkitItemBuilder.prototype.amount = function (amount) {
        this.itemStack.setAmount(amount);
        return this;
    };
    BukkitItemBuilder.prototype.durability = function (durability) {
        this.itemStack.setDurability(durability);
        return this;
    };
    BukkitItemBuilder.prototype.clone = function () {
        return this.build().clone();
    };
    BukkitItemBuilder.prototype.build = function () {
        this.itemStack.setItemMeta(this.itemMeta);
        return this.itemStack;
    };
    return BukkitItemBuilder;
}());
exports.BukkitItemBuilder = BukkitItemBuilder;
