// @ccms/bukkit/dist/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BukkitNativePluginManager = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var Bukkit = Java.type('org.bukkit.Bukkit');
var BukkitNativePluginManager = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitNativePluginManager, _super);
    function BukkitNativePluginManager() {
        var _this = _super.call(this) || this;
        _this.bukkitPluginManager = Bukkit.getPluginManager();
        return _this;
    }
    BukkitNativePluginManager.prototype.list = function () {
        var _this = this;
        return Java.from(this.bukkitPluginManager.getPlugins()).map(function (plugin) { return _this.convert(plugin); });
    };
    BukkitNativePluginManager.prototype.has = function (name) {
        return !!this.bukkitPluginManager.getPlugin(name);
    };
    BukkitNativePluginManager.prototype.get = function (name) {
        return this.convert(this.bukkitPluginManager.getPlugin(name));
    };
    BukkitNativePluginManager.prototype.enable = function (name) {
        var origin = this.bukkitPluginManager.getPlugin(name);
        if (!origin) {
            throw new Error("Native Plugin ".concat(name, " not found."));
        }
        if (!origin.isEnabled()) {
            this.bukkitPluginManager.enablePlugin(origin);
        }
        return this.convert(origin);
    };
    BukkitNativePluginManager.prototype.disable = function (name) {
        var origin = this.bukkitPluginManager.getPlugin(name);
        if (!origin) {
            throw new Error("Native Plugin ".concat(name, " not found."));
        }
        if (origin.isEnabled()) {
            this.bukkitPluginManager.disablePlugin(origin);
        }
        return this.convert(origin);
    };
    BukkitNativePluginManager.prototype.convert = function (plugin) {
        if (!plugin)
            return plugin;
        var desc = plugin.getDescription();
        return {
            name: plugin.getName(),
            version: desc.getVersion(),
            authors: Java.from(desc.getAuthors()),
            depends: Java.from(desc.getDepend()),
            softDepends: Java.from(desc.getSoftDepend()),
            enable: plugin.isEnabled(),
            origin: plugin
        };
    };
    return BukkitNativePluginManager;
}(api_1.server.NativePluginManager));
exports.BukkitNativePluginManager = BukkitNativePluginManager;
