// @ccms/bukkit/dist/server.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BukkitServer = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var reflect = require("@ccms/common/dist/reflect");
var Bukkit = Java.type('org.bukkit.Bukkit');
var BukkitServer = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitServer, _super);
    function BukkitServer() {
        var _this = _super.call(this) || this;
        _this.pluginsFolder = Bukkit.getUpdateFolderFile().getParentFile().getCanonicalPath();
        return _this;
    }
    BukkitServer.prototype.getPlayer = function (name) {
        return Bukkit.getPlayerExact(name);
    };
    BukkitServer.prototype.getVersion = function () {
        return Bukkit.getVersion();
    };
    BukkitServer.prototype.getOnlinePlayers = function () {
        return Bukkit.getOnlinePlayers();
    };
    BukkitServer.prototype.getConsoleSender = function () {
        return Bukkit.getConsoleSender();
    };
    BukkitServer.prototype.getService = function (service) {
        var _a;
        return (_a = Bukkit.getServicesManager().getRegistration(base.getClass(service))) === null || _a === void 0 ? void 0 : _a.getProvider();
    };
    BukkitServer.prototype.broadcast = function (message, permission) {
        return Bukkit.broadcast(message, permission);
    };
    BukkitServer.prototype.broadcastMessage = function (message) {
        return Bukkit.broadcastMessage(message);
    };
    BukkitServer.prototype.dispatchCommand = function (sender, command) {
        if (typeof sender === 'string') {
            sender = this.getPlayer(sender);
        }
        return Bukkit.dispatchCommand(sender, command);
    };
    BukkitServer.prototype.dispatchConsoleCommand = function (command) {
        return Bukkit.dispatchCommand(Bukkit.getConsoleSender(), command);
    };
    BukkitServer.prototype.getPluginsFolder = function () {
        return this.pluginsFolder;
    };
    BukkitServer.prototype.getNativePluginManager = function () {
        return Bukkit.getPluginManager();
    };
    BukkitServer.prototype.getDedicatedServer = function () {
        return reflect.on(Bukkit.getServer()).get('console').get();
    };
    BukkitServer.prototype.getNettyPipeline = function () {
        return this.pipeline;
    };
    BukkitServer.prototype.getRootLogger = function () {
        return this.rootLogger;
    };
    BukkitServer = tslib_1.__decorate([
        (0, container_1.provideSingleton)(api_1.server.Server),
        tslib_1.__metadata("design:paramtypes", [])
    ], BukkitServer);
    return BukkitServer;
}(api_1.server.ReflectServer));
exports.BukkitServer = BukkitServer;
