// @ccms/bukkit/dist/task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BukkitTask = exports.BukkitTaskManager = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var Bukkit = Java.type('org.bukkit.Bukkit');
var BukkitRunnable = Java.type('org.bukkit.scheduler.BukkitRunnable');
var Callable = Java.type('java.util.concurrent.Callable');
var BukkitTaskManager = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitTaskManager, _super);
    function BukkitTaskManager() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BukkitTaskManager.prototype.create0 = function (owner, func, id) {
        return new BukkitTask(owner, func, id);
    };
    BukkitTaskManager.prototype.callSyncMethod = function (func) {
        return Bukkit.getScheduler().callSyncMethod(base.getInstance(), new Callable({ call: function () { return func(); } })).get();
    };
    BukkitTaskManager.prototype.disable0 = function () {
        Bukkit.getScheduler().cancelTasks(base.getInstance());
    };
    BukkitTaskManager = tslib_1.__decorate([
        (0, container_1.provideSingleton)(api_1.task.TaskManager)
    ], BukkitTaskManager);
    return BukkitTaskManager;
}(api_1.task.TaskManager));
exports.BukkitTaskManager = BukkitTaskManager;
var BukkitTask = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitTask, _super);
    function BukkitTask() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BukkitTask.prototype.submit0 = function () {
        var _this = this;
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var run = new BukkitRunnable({ run: function () { return _this.run.apply(_this, tslib_1.__spreadArray([], tslib_1.__read(args), false)); } });
        var suffix = this.isAsync ? 'Asynchronously' : '';
        if (this.interval) {
            return run["runTaskTimer".concat(suffix)](base.getInstance(), this.laterTime, this.interval);
        }
        else {
            return run["runTaskLater".concat(suffix)](base.getInstance(), this.laterTime);
        }
    };
    return BukkitTask;
}(api_1.task.Task));
exports.BukkitTask = BukkitTask;
