// @ccms/common/dist/fs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.exists = exports.del = exports.move = exports.list = exports.save = exports.read = exports.copy = exports.path = exports.create = exports.mkdirs = exports.file = exports.concat = void 0;
require("@ccms/nashorn");
/*global Java, base, module, exports, require, __FILE__*/
var Path = Java.type("java.nio.file.Path");
var JavaString = Java.type("java.lang.String");
var File = Java.type("java.io.File");
var Files = Java.type("java.nio.file.Files");
var Collector = Java.type("java.util.stream.Collector");
var separatorChar = File.separatorChar;
var StandardCopyOption = Java.type("java.nio.file.StandardCopyOption");
/**
 * 用文件分割符合并路径
 */
function concat() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return args.join(separatorChar);
}
exports.concat = concat;
/**
 * 获得文件
 * @constructor(file)
 * @constructor(dir,file)
 * @returns {*}
 */
function file() {
    var opts = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        opts[_i] = arguments[_i];
    }
    if (!opts[0]) {
        throw new Error("文件名称不得为 undefined 或者 null !");
    }
    switch (opts.length) {
        case 1:
            var f = opts[0];
            if (f instanceof File) {
                return f;
            }
            if (typeof f === "string") {
                return new File(f);
            }
            if (f instanceof Path) {
                return f.toFile();
            }
            break;
        default:
            return new File(file(opts[0]), opts[1]);
    }
}
exports.file = file;
/**
 * 创建目录
 * @param path
 */
function mkdirs(path) {
    // noinspection JSUnresolvedVariable
    file(path).parentFile.mkdirs();
}
exports.mkdirs = mkdirs;
/**
 * 创建文件
 * @param file
 */
function create(path) {
    var f = file(path);
    if (!f.exists()) {
        mkdirs(f);
        f.createNewFile();
    }
}
exports.create = create;
/**
 * 获得文件规范路径
 * @param file
 * @returns {*}
 */
function path(f) {
    return file(f).canonicalPath;
}
exports.path = path;
/**
 * 复制文件
 * @param inputStream 输入流
 * @param target 目标文件
 * @param override 是否覆盖
 */
function copy(inputStream, target, override) {
    Files.copy(inputStream, target.toPath(), StandardCopyOption[override ? 'REPLACE_EXISTING' : 'ATOMIC_MOVE']);
}
exports.copy = copy;
/**
 * 读取文件
 * @param path 文件路径
 */
function read(path) {
    var file = file(path);
    if (!file.exists()) {
        console.warn('读取文件', file, '错误 文件不存在!');
        return;
    }
    // noinspection JSPrimitiveTypeWrapperUsage
    return new JavaString(Files.readAllBytes(file.toPath()), "UTF-8");
}
exports.read = read;
/**
 * 保存内容文件
 * @param path 路径
 * @param content 内容
 * @param override 是否覆盖
 */
function save(path, content, override) {
    var file = file(path);
    if (file.parentFile) {
        file.parentFile.mkdirs();
    }
    Files.write(file.toPath(), new JavaString(content).getBytes("UTF-8"));
}
exports.save = save;
/**
 * 列出目录文件
 * @param path
 */
function list(path) {
    var dir = file(path);
    // noinspection JSValidateTypes
    if (dir.isDirectory()) {
        return Files.list(dir.toPath());
    }
    console.debug('路径', path, '不是一个目录 返回空数组!');
    return [];
}
exports.list = list;
/**
 * 移动文件
 * @param src 原始目录
 * @param des 目标目录
 * @param override 是否覆盖
 */
function move(src, des, override) {
    Files.move(file(src).toPath(), file(des).toPath(), override ? StandardCopyOption['REPLACE_EXISTING'] : StandardCopyOption['ATOMIC_MOVE']);
}
exports.move = move;
function del(file) {
    file = file(file);
    if (!file.exists()) {
        return;
    }
    if (file.isDirectory()) {
        Files.list(file.toPath()).collect(Collector.toList()).forEach(function (f) {
            del(f);
        });
    }
    Files.delete(file.toPath());
}
exports.del = del;
function exists(f) {
    return file(f).exists();
}
exports.exists = exists;
