// @ccms/common/dist/reflect.js
"use strict";
/// <reference types="@javatypes/jdk" />
var tslib_1 = require("tslib");
var JavaClass = Java.type('java.lang.Class');
var JavaObject = Java.type('java.lang.Object');
var NoSuchFieldException = Java.type('java.lang.NoSuchFieldException');
var fieldCache = new Map();
var methodCache = new Map();
/**
 * 反射工具类
 * Created by MiaoWoo on 2017/2/9 0009.
 */
var Reflect = /** @class */ (function () {
    function Reflect(obj) {
        // if (obj === undefined || obj === null) { throw Error(`reflect object can't be ${obj}!`) }
        if (obj instanceof JavaClass) {
            this.obj = null;
            this.class = obj;
        }
        else {
            this.obj = obj;
            if (obj !== null && obj !== undefined && obj.class) {
                this.class = obj.class;
            }
        }
    }
    Reflect.prototype.method = function (name) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return declaredMethod.apply(void 0, tslib_1.__spreadArray([this.class, name], tslib_1.__read(args), false));
    };
    Reflect.prototype.methods = function () {
        return Java.from(declaredMethods(this.class));
    };
    Reflect.prototype.field = function (nameOrIndex, declared) {
        if (declared === void 0) { declared = false; }
        if (nameOrIndex == undefined || nameOrIndex == null)
            throw new Error("reflect field name can't be ".concat(nameOrIndex, " from ").concat(this.class.getName(), "!"));
        var key = this.class.getName() + ':' + nameOrIndex + ':' + declared;
        if (fieldCache.has(key)) {
            return fieldCache.get(key);
        }
        var field = null;
        if (typeof nameOrIndex == "number") {
            field = this.fields(declared)[nameOrIndex];
        }
        else {
            try {
                // Try getting a public field
                field = this.class.getField(nameOrIndex);
            }
            catch (ex) {
                // Try again, getting a non-public field
                field = declaredField(this.class, nameOrIndex);
            }
        }
        if (!field)
            throw new Error("can't reflect field ".concat(typeof nameOrIndex == "number" ? 'index' : 'name', " ").concat(nameOrIndex, " from ").concat(this.class.getName(), "!"));
        fieldCache.set(key, field);
        return accessible(field);
    };
    Reflect.prototype.fields = function (declared) {
        if (declared === void 0) { declared = false; }
        return Java.from(declared ? this.class.getDeclaredFields() : this.class.getFields());
    };
    Reflect.prototype.values = function (declared) {
        var _this = this;
        if (declared === void 0) { declared = false; }
        return this.fields(declared).reduce(function (cache, field) { return cache[field.getName()] = _this.get(field.getName()).get(); }, {});
    };
    Reflect.prototype.call = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        var params = args.slice(1);
        var method = accessible(declaredMethod.apply(void 0, tslib_1.__spreadArray([this.class, args[0]], tslib_1.__read(types(params)), false)));
        var result = method.invoke(this.get(), params);
        return result && on(result);
    };
    Reflect.prototype.get = function (param, declared) {
        if (declared === void 0) { declared = true; }
        if (arguments.length === 0)
            return this.obj;
        return on(this.field(param, declared).get(this.obj));
    };
    Reflect.prototype.set = function (param, value, declared) {
        if (declared === void 0) { declared = true; }
        this.field(param, declared).set(this.obj, value);
        return this;
    };
    Reflect.prototype.create = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return on(declaredConstructor(this.class, args).newInstance(args));
    };
    return Reflect;
}());
/**
 * Get an array of types for an array of objects
 */
function types(values, def) {
    if (values === null) {
        return [];
    }
    var result = [];
    values.forEach(function (t) { return result.push((t || def) ? JavaObject.class : t instanceof JavaClass ? t : t.class); });
    return result;
}
function accessible(accessible) {
    if (!(accessible === null || accessible === void 0 ? void 0 : accessible.isAccessible())) {
        accessible.setAccessible(true);
    }
    return accessible;
}
function declaredConstructor(clazz, param) {
    var constructor;
    try {
        constructor = clazz.getDeclaredConstructor(types(param));
    }
    catch (ex) {
        try {
            constructor = clazz.getDeclaredConstructor(types(param, true));
        }
        catch (ex) {
            constructor = clazz.getDeclaredConstructors()[0];
        }
    }
    return accessible(constructor);
}
function declaredField(clazz, name) {
    if (!clazz) {
        throw Error("target class can't be ".concat(clazz, "!"));
    }
    var target = clazz;
    var field = null;
    // noinspection JSUnresolvedVariable
    while (target !== JavaObject.class) {
        console.debug("reflect field ".concat(name, " from ").concat(target.getName()));
        try {
            field = target.getDeclaredField(name);
            if (field !== null) {
                break;
            }
        }
        catch (error) {
            if (target === undefined) {
                break;
            }
            target = target.getSuperclass();
            console.debug("\u5207\u6362\u5230\u8D85\u7C7B: ".concat(target.getName()));
        }
    }
    if (field === null) {
        throw new NoSuchFieldException(name + " is not found in " + clazz.getName());
    }
    return field;
}
function declaredMethod(clazz, nameOrIndex) {
    var e_1, _a;
    var clazzs = [];
    for (var _i = 2; _i < arguments.length; _i++) {
        clazzs[_i - 2] = arguments[_i];
    }
    var key = clazz.getName() + '.' + nameOrIndex + ':' + (clazzs || []).map(function (c) { return c.getName(); }).join(':');
    if (methodCache.has(key)) {
        return methodCache.get(key);
    }
    var target = clazz;
    if (typeof nameOrIndex === "number") {
        methodCache.set(key, declaredMethods(clazz)[nameOrIndex]);
    }
    else {
        while (target !== JavaObject.class && !methodCache.has(key)) {
            try {
                console.debug("reflect method ".concat(typeof nameOrIndex == "number" ? 'index' : 'name', " ").concat(nameOrIndex, " from ").concat(target.getName()));
                try {
                    methodCache.set(key, target.getMethod(nameOrIndex, clazzs));
                }
                catch (ex) {
                    try {
                        methodCache.set(key, target.getDeclaredMethod(nameOrIndex, clazzs));
                    }
                    catch (ex) {
                        try {
                            for (var _b = (e_1 = void 0, tslib_1.__values(Java.from(declaredMethods(target)))), _c = _b.next(); !_c.done; _c = _b.next()) {
                                var m = _c.value;
                                if (m.getName() == nameOrIndex) {
                                    methodCache.set(key, m);
                                    break;
                                }
                            }
                        }
                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                        finally {
                            try {
                                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                            }
                            finally { if (e_1) throw e_1.error; }
                        }
                        throw new Error("method ".concat(typeof nameOrIndex == "number" ? 'index' : 'name', " ").concat(nameOrIndex, " not found."));
                    }
                }
            }
            catch (error) {
                if (target === undefined) {
                    break;
                }
                target = target.getSuperclass();
                console.debug("\u5207\u6362\u5230\u8D85\u7C7B: ".concat(target.getName()));
            }
        }
    }
    if (!methodCache.has(key))
        throw new Error("can't reflect method ".concat(typeof nameOrIndex == "number" ? 'index' : 'name', " ").concat(nameOrIndex, " from ").concat(clazz.getName(), "!"));
    return methodCache.get(key);
}
function declaredMethods(clazz) {
    return clazz.getDeclaredMethods();
}
function mapToObject(javaObj) {
    if (!Java.isJavaObject(javaObj)) {
        throw new TypeError("\u53C2\u6570 ".concat(javaObj, " \u4E0D\u662F\u4E00\u4E2AJava\u5BF9\u8C61!"));
    }
    var target = Proxy.newProxy(javaObj, {
        apply: function (target, name, args) { return args ? javaObj[name](args) : javaObj[name](); }
    });
    return target;
}
function on(obj) {
    // if (!obj || !obj.class) { throw new TypeError(`参数 ${obj} 不是一个Java对象!`) }
    return new Reflect(obj);
}
module.exports = {
    on: on,
    accessible: accessible,
    declaredMethods: declaredMethods,
    mapToObject: mapToObject
};
