// @ccms/common/dist/tellraw.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ChatMessagePart = /** @class */ (function () {
    function ChatMessagePart() {
        this.internal = {};
    }
    Object.defineProperty(ChatMessagePart.prototype, "text", {
        get: function () {
            return this.internal.text;
        },
        set: function (text) {
            this.internal.text = text;
        },
        enumerable: false,
        configurable: true
    });
    ChatMessagePart.prototype.click = function (action, value) {
        this.internal.clickEvent = { action: action, value: value };
    };
    ChatMessagePart.prototype.hover = function (action, value) {
        this.internal.hoverEvent = { action: action, value: value };
    };
    ChatMessagePart.prototype.convert = function () {
        return this.internal;
    };
    return ChatMessagePart;
}());
var Tellraw = /** @class */ (function () {
    function Tellraw() {
        this.cache = '';
        this.parts = [new ChatMessagePart()];
    }
    Tellraw.create = function () {
        return new Tellraw().then(Tellraw.duplicateChar);
    };
    Tellraw.prototype.then = function (part) {
        if (typeof part === "string") {
            var newPart = new ChatMessagePart();
            newPart.text = part;
            this.then(newPart);
            return this;
        }
        var last = this.latest();
        if (!last.text) {
            last.text = part.text;
        }
        else {
            this.parts.push(part);
        }
        this.cache = null;
    };
    Tellraw.prototype.text = function (text) {
        this.latest().text = text;
        return this;
    };
    Tellraw.prototype.tip = function (texts) {
        return this.hover(texts);
    };
    Tellraw.prototype.hover = function (texts) {
        this.latest().hover("show_text", texts);
        return this;
    };
    Tellraw.prototype.item = function (item) {
        this.latest().hover("show_item", item);
        return this;
    };
    Tellraw.prototype.command = function (command) {
        this.latest().click("run_command", command);
        return this;
    };
    Tellraw.prototype.suggest = function (url) {
        this.latest().click("suggest_command", url);
        return this;
    };
    Tellraw.prototype.file = function (path) {
        this.latest().click("open_file", path);
        return this;
    };
    Tellraw.prototype.link = function (url) {
        this.latest().click("open_url", url);
        return this;
    };
    Tellraw.prototype.latest = function () {
        return this.parts[this.parts.length - 1];
    };
    Tellraw.prototype.json = function () {
        if (!this.cache) {
            var temp = [];
            this.parts.forEach(function (t) {
                temp.push(t.convert());
            });
            this.cache = JSON.stringify(temp);
            console.trace(this.cache);
        }
        return this.cache;
    };
    Tellraw.prototype.string = function () {
        var temp = '';
        this.parts.forEach(function (t) {
            temp += t.text;
        });
        return temp;
    };
    Tellraw.duplicateChar = '§卐';
    return Tellraw;
}());
exports.default = Tellraw;
