// @ccms/plugin/dist/command.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginCommandManager = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var utils_1 = require("./utils");
var PluginCommandManager = /** @class */ (function () {
    function PluginCommandManager() {
        process.on('plugin.before.enable', this.registryCommand.bind(this));
        process.on('plugin.after.disable', this.unregistryCommand.bind(this));
    }
    PluginCommandManager_1 = PluginCommandManager;
    PluginCommandManager.prototype.registryCommand = function (pluginInstance) {
        var e_1, _a, e_2, _b;
        var cmds = (0, utils_1.getPluginCommandMetadata)(pluginInstance);
        var tabs = (0, utils_1.getPluginTabCompleterMetadata)(pluginInstance);
        try {
            for (var cmds_1 = tslib_1.__values(cmds), cmds_1_1 = cmds_1.next(); !cmds_1_1.done; cmds_1_1 = cmds_1.next()) {
                var cmd = cmds_1_1.value;
                if (!this.ServerChecker.check(cmd.servers)) {
                    console.debug("[".concat(pluginInstance.description.name, "] ").concat(cmd.target.constructor.name, " incompatible command ").concat(cmd.name, " server(").concat(cmd.servers, ") ignore."));
                    continue;
                }
                try {
                    for (var _c = (e_2 = void 0, tslib_1.__values(tslib_1.__spreadArray([cmd.name], tslib_1.__read(cmd.alias), false))), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var command_1 = _d.value;
                        var _e = tslib_1.__read(this.generateAutoMainCommand(pluginInstance, cmd, tabs.get(command_1)), 2), cmdExecutor = _e[0], cmdCompleter = _e[1];
                        this.CommandManager.on(pluginInstance, command_1, {
                            cmd: cmdExecutor.bind(pluginInstance),
                            tab: cmdCompleter === null || cmdCompleter === void 0 ? void 0 : cmdCompleter.bind(pluginInstance)
                        });
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (cmds_1_1 && !cmds_1_1.done && (_a = cmds_1.return)) _a.call(cmds_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    PluginCommandManager.prototype.unregistryCommand = function (pluginInstance) {
        var e_3, _a, e_4, _b;
        var cmds = (0, utils_1.getPluginCommandMetadata)(pluginInstance);
        try {
            for (var cmds_2 = tslib_1.__values(cmds), cmds_2_1 = cmds_2.next(); !cmds_2_1.done; cmds_2_1 = cmds_2.next()) {
                var cmd = cmds_2_1.value;
                if (!this.ServerChecker.check(cmd.servers)) {
                    console.debug("[".concat(pluginInstance.description.name, "] ").concat(cmd.target.constructor.name, " incompatible command ").concat(cmd.name, " server(").concat(cmd.servers, ") ignore."));
                    continue;
                }
                try {
                    for (var _c = (e_4 = void 0, tslib_1.__values(tslib_1.__spreadArray([cmd.name], tslib_1.__read(cmd.alias), false))), _d = _c.next(); !_d.done; _d = _c.next()) {
                        var command_2 = _d.value;
                        this.CommandManager.off(pluginInstance, command_2);
                    }
                }
                catch (e_4_1) { e_4 = { error: e_4_1 }; }
                finally {
                    try {
                        if (_d && !_d.done && (_b = _c.return)) _b.call(_c);
                    }
                    finally { if (e_4) throw e_4.error; }
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (cmds_2_1 && !cmds_2_1.done && (_a = cmds_2.return)) _a.call(cmds_2);
            }
            finally { if (e_3) throw e_3.error; }
        }
    };
    PluginCommandManager.prototype.generateAutoMainCommand = function (pluginInstance, cmd, tab) {
        var cmdExecutor = pluginInstance[cmd.executor];
        var cmdCompleter = tab ? pluginInstance[tab.executor] : undefined;
        var cmdSubCache = Object.keys(pluginInstance.constructor.prototype).filter(function (s) { return s.startsWith('cmd'); }).map(function (s) { return s.substring(3); });
        if (cmd.autoMain) {
            cmdExecutor = function (sender, command, args) {
                var subcommand = args[0];
                var cmdKey = 'cmd' + subcommand;
                if (!cmdSubCache.includes(subcommand)) {
                    if (!pluginInstance[cmd.executor].apply(pluginInstance, [sender, command, args])) {
                        subcommand && pluginInstance.logger.sender(sender, "\u00A74\u672A\u77E5\u7684\u547D\u4EE4: \u00A7b/".concat(command, " \u00A7c").concat(subcommand));
                        pluginInstance.logger.sender(sender, pluginInstance['cmdhelp'] ?
                            "\u00A76\u8BF7\u6267\u884C \u00A7b/".concat(command, " \u00A7ahelp \u00A76\u67E5\u770B\u5E2E\u52A9!") :
                            [
                                "\u00A76\u63D2\u4EF6: \u00A7b".concat(pluginInstance.description.name),
                                "\u00A76\u7248\u672C: \u00A7a".concat(pluginInstance.description.version)
                            ]);
                    }
                    return;
                }
                var subcommandexec = pluginInstance[cmdKey];
                var permission;
                if (cmd.permission && sender.hasPermission) {
                    if (typeof cmd.permission == "string") {
                        permission = cmd.permission;
                    }
                    else {
                        permission = "".concat(pluginInstance.description.name.toLocaleLowerCase(), ".").concat(command, ".").concat(subcommand || 'main');
                    }
                    if (!sender.hasPermission(permission)) {
                        return pluginInstance.logger.sender(sender, "\u00A7c\u4F60\u9700\u8981 \u00A74".concat(permission, " \u00A7c\u6743\u9650 \u624D\u53EF\u6267\u884C\u6B64\u547D\u4EE4."));
                    }
                }
                args.shift();
                return subcommandexec.apply(pluginInstance, tslib_1.__spreadArray([sender], tslib_1.__read(args), false));
            };
            var originCompleter_1 = cmdCompleter;
            cmdCompleter = function (sender, command, args) {
                var permission;
                if (cmd.permission && sender.hasPermission) {
                    if (typeof cmd.permission == "string") {
                        permission = cmd.permission;
                    }
                    else {
                        permission = "".concat(pluginInstance.description.name.toLocaleLowerCase(), ".").concat(command);
                    }
                    if (!sender.hasPermission(permission)) {
                        return [];
                    }
                }
                return (args.length == 1 ? cmdSubCache : []).concat((originCompleter_1 === null || originCompleter_1 === void 0 ? void 0 : originCompleter_1.apply(pluginInstance, [sender, command, args])) || []);
            };
        }
        if (!cmdCompleter) {
            console.debug("[".concat(pluginInstance.description.name, "] command ").concat(cmd.name, " is not registry tabCompleter"));
        }
        return [cmdExecutor, cmdCompleter];
    };
    var PluginCommandManager_1;
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", api_1.command.Command)
    ], PluginCommandManager.prototype, "CommandManager", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", api_1.server.ServerChecker)
    ], PluginCommandManager.prototype, "ServerChecker", void 0);
    PluginCommandManager = PluginCommandManager_1 = tslib_1.__decorate([
        (0, container_1.provideSingleton)(PluginCommandManager_1),
        tslib_1.__metadata("design:paramtypes", [])
    ], PluginCommandManager);
    return PluginCommandManager;
}());
exports.PluginCommandManager = PluginCommandManager;
