// @ccms/plugin/dist/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginConfigManager = void 0;
var tslib_1 = require("tslib");
var container_1 = require("@ccms/container");
var fs = require("@ccms/common/dist/fs");
var utils_1 = require("./utils");
var interfaces_1 = require("./config/interfaces");
require("./config/loader");
var PluginConfigManager = /** @class */ (function () {
    function PluginConfigManager() {
        this.configLoaderMap = new Map();
        process.on('plugin.before.load', this.loadConfig.bind(this));
        process.on('plugin.after.disable', this.saveConfig.bind(this));
    }
    PluginConfigManager_1 = PluginConfigManager;
    PluginConfigManager.prototype.initialize = function () {
        var _this = this;
        var configLoader = this.container.getAll(interfaces_1.PluginConfigLoader);
        configLoader.forEach(function (scanner) {
            console.debug("loading config loader ".concat(scanner.type, "..."));
            _this.configLoaderMap.set(scanner.type, scanner);
        });
    };
    PluginConfigManager.prototype.getConfigLoader = function (format) {
        if (!this.configLoaderMap.has(format)) {
            throw new Error("Unsupport config format ".concat(format, " !"));
        }
        return this.configLoaderMap.get(format);
    };
    PluginConfigManager.prototype.loadConfig = function (plugin) {
        var e_1, _a;
        var configs = (0, utils_1.getPluginConfigMetadata)(plugin);
        try {
            for (var configs_1 = tslib_1.__values(configs), configs_1_1 = configs_1.next(); !configs_1_1.done; configs_1_1 = configs_1.next()) {
                var _b = tslib_1.__read(configs_1_1.value, 2), _ = _b[0], config = _b[1];
                this.loadConfig0(plugin, config);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (configs_1_1 && !configs_1_1.done && (_a = configs_1.return)) _a.call(configs_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    PluginConfigManager.prototype.saveConfig = function (plugin) {
        var e_2, _a;
        var configs = (0, utils_1.getPluginConfigMetadata)(plugin);
        try {
            for (var configs_2 = tslib_1.__values(configs), configs_2_1 = configs_2.next(); !configs_2_1.done; configs_2_1 = configs_2.next()) {
                var _b = tslib_1.__read(configs_2_1.value, 2), _ = _b[0], config = _b[1];
                config.autosave && this.saveConfig0(plugin, config);
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (configs_2_1 && !configs_2_1.done && (_a = configs_2.return)) _a.call(configs_2);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    PluginConfigManager.prototype.createConfig = function (plugin, metadata, value) {
        var _this = this;
        Object.defineProperties(value, {
            'save': { value: function () { return _this.saveConfig0(plugin, metadata); } },
            'reload': { value: function () { return _this.loadConfig0(plugin, metadata); } }
        });
        return value;
    };
    PluginConfigManager.prototype.defienConfigProp = function (plugin, metadata, value) {
        Object.defineProperty(plugin, metadata.variable, {
            value: this.createConfig(plugin, metadata, value),
            configurable: true
        });
    };
    PluginConfigManager.prototype.loadConfig0 = function (plugin, metadata) {
        var _a;
        try {
            var defaultValue = (_a = metadata.default) !== null && _a !== void 0 ? _a : plugin[metadata.variable];
            metadata.file = fs.concat(fs.file(plugin.description.loadMetadata.file).parent, plugin.description.name, metadata.filename);
            var configLoader = this.getConfigLoader(metadata.format);
            if (!fs.exists(metadata.file) && defaultValue) {
                base.save(metadata.file, configLoader.dump(defaultValue));
                console.i18n("ms.plugin.manager.config.save.default", {
                    plugin: plugin.description.name,
                    name: metadata.name,
                    format: metadata.format
                });
            }
            var configValue = configLoader.load(base.read(metadata.file)) || {};
            if (metadata.migrate && defaultValue && this.setDefaultValue(configValue, defaultValue, !!metadata.default)) {
                base.save(metadata.file, configLoader.dump(configValue));
            }
            console.debug("[".concat(plugin.description.name, "] Load Config ").concat(metadata.variable, " from file ").concat(metadata.file));
            this.defienConfigProp(plugin, metadata, configValue);
        }
        catch (error) {
            console.i18n("ms.plugin.manager.config.load.error", {
                plugin: plugin.description.name,
                name: metadata.name,
                format: metadata.format,
                error: error
            });
            console.ex(error);
        }
    };
    PluginConfigManager.prototype.setDefaultValue = function (configValue, defaultValue, deepCopy) {
        var e_3, _a;
        var needSave = false;
        try {
            for (var _b = tslib_1.__values(Object.keys(defaultValue)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var key = _c.value;
                // 当配置文件不存在当前属性时才进行赋值
                if (!Object.prototype.hasOwnProperty.call(configValue, key)) {
                    configValue[key] = defaultValue[key];
                    needSave = true;
                }
                else if (Object.prototype.toString.call(configValue[key]) == "[object Object]" && deepCopy) {
                    // 对象需要递归检测
                    needSave || (needSave = this.setDefaultValue(configValue[key], defaultValue[key], deepCopy));
                }
            }
        }
        catch (e_3_1) { e_3 = { error: e_3_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_3) throw e_3.error; }
        }
        return needSave;
    };
    PluginConfigManager.prototype.saveConfig0 = function (plugin, metadata) {
        try {
            metadata.file = fs.concat(fs.file(plugin.description.loadMetadata.file).parent, plugin.description.name, metadata.filename);
            var result = this.getConfigLoader(metadata.format).dump(plugin[metadata.variable]);
            base.save(metadata.file, result);
            console.debug("[".concat(plugin.description.name, "] Save Config ").concat(metadata.variable, " to file ").concat(metadata.file, " =>\n").concat(result.substring(0, 500)));
            return true;
        }
        catch (error) {
            console.i18n("ms.plugin.manager.config.save.error", {
                plugin: plugin.description.name,
                name: metadata.name,
                format: metadata.format,
                error: error
            });
            console.ex(error);
            return false;
        }
    };
    var PluginConfigManager_1;
    tslib_1.__decorate([
        (0, container_1.Autowired)(container_1.ContainerInstance),
        tslib_1.__metadata("design:type", container_1.Container)
    ], PluginConfigManager.prototype, "container", void 0);
    tslib_1.__decorate([
        (0, container_1.postConstruct)(),
        tslib_1.__metadata("design:type", Function),
        tslib_1.__metadata("design:paramtypes", []),
        tslib_1.__metadata("design:returntype", void 0)
    ], PluginConfigManager.prototype, "initialize", null);
    PluginConfigManager = PluginConfigManager_1 = tslib_1.__decorate([
        (0, container_1.provideSingleton)(PluginConfigManager_1),
        tslib_1.__metadata("design:paramtypes", [])
    ], PluginConfigManager);
    return PluginConfigManager;
}());
exports.PluginConfigManager = PluginConfigManager;
