// @ccms/plugin/dist/decorators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disable = exports.enable = exports.load = exports.playerdata = exports.config = exports.listener = exports.tab = exports.cmd = exports.plugin = void 0;
var tslib_1 = require("tslib");
var container_1 = require("@ccms/container");
var constants_1 = require("./constants");
var utils_1 = require("./utils");
/**
 * MiaoScript plugin
 * @param metadata PluginMetadata
 */
function plugin(metadata) {
    return function (target) {
        metadata = tslib_1.__assign({ name: target.name, version: '1.0.0', author: 'Unknow', target: target, type: 'ioc' }, metadata);
        (0, container_1.decorate)((0, container_1.injectable)(), target);
        Reflect.defineMetadata(constants_1.METADATA_KEY.plugin, metadata, target);
        var previousMetadata = (0, utils_1.getPluginMetadatas)();
        previousMetadata.set(metadata.name, metadata);
        Reflect.defineMetadata(constants_1.METADATA_KEY.plugin, previousMetadata, Reflect);
        var previousSources = (0, utils_1.getPluginSources)();
        previousSources.set(metadata.source.toString(), metadata);
        Reflect.defineMetadata(constants_1.METADATA_KEY.source, previousSources, Reflect);
    };
}
exports.plugin = plugin;
/**
 * MiaoScript Command
 * @param metadata CommandMetadata
 */
function cmd(metadata) {
    if (metadata === void 0) { metadata = {}; }
    return function (target, key, value) {
        metadata.name = metadata.name || key;
        metadata.alias = metadata.alias || [];
        metadata.target = target;
        metadata.executor = key;
        metadata.paramtypes = Reflect.getMetadata("design:paramtypes", target, key);
        Reflect.defineMetadata(constants_1.METADATA_KEY.cmd, tslib_1.__spreadArray([metadata], tslib_1.__read((0, utils_1.getPluginCommandMetadata)(target)), false), target.constructor);
    };
}
exports.cmd = cmd;
/**
 * MiaoScript TabComplete
 * @param metadata TabCompleterMetadata
 */
function tab(metadata) {
    if (metadata === void 0) { metadata = {}; }
    return function (target, key, value) {
        var _a;
        metadata.name = metadata.name || (key.startsWith('tab') ? key.split('tab', 2)[1] : key);
        if (!metadata.name) {
            return;
        }
        metadata.alias = metadata.alias || [];
        metadata.target = target;
        metadata.executor = key;
        metadata.paramtypes = Reflect.getMetadata("design:paramtypes", target, key);
        var previousMetadata = (0, utils_1.getPluginTabCompleterMetadata)(target);
        previousMetadata.set(metadata.name, metadata);
        (_a = metadata.alias) === null || _a === void 0 ? void 0 : _a.forEach(function (name) { return previousMetadata.set(name, metadata); });
        Reflect.defineMetadata(constants_1.METADATA_KEY.tab, previousMetadata, target.constructor);
    };
}
exports.tab = tab;
/**
 * MiaoScript Listener
 * @param metadata ListenerMetadata
 */
function listener(metadata) {
    if (metadata === void 0) { metadata = {}; }
    return function (target, key, value) {
        metadata.name = metadata.name || key;
        metadata.target = target;
        metadata.executor = key;
        Reflect.defineMetadata(constants_1.METADATA_KEY.listener, tslib_1.__spreadArray([metadata], tslib_1.__read((0, utils_1.getPluginListenerMetadata)(target)), false), target.constructor);
        Reflect.defineMetadata(constants_1.METADATA_KEY.listener, metadata, target[key]);
    };
}
exports.listener = listener;
function config(metadata) {
    if (metadata === void 0) { metadata = {}; }
    return function (target, key) {
        var _a, _b, _c, _d, _e;
        metadata.name = metadata.name || key;
        metadata.variable = key;
        metadata.version = (_a = metadata.version) !== null && _a !== void 0 ? _a : 1;
        metadata.format = (_b = metadata.format) !== null && _b !== void 0 ? _b : 'yml';
        metadata.migrate = (_c = metadata.migrate) !== null && _c !== void 0 ? _c : true;
        metadata.autosave = (_d = metadata.autosave) !== null && _d !== void 0 ? _d : false;
        metadata.filename = (_e = metadata.filename) !== null && _e !== void 0 ? _e : metadata.name + '.' + metadata.format;
        var previousMetadata = (0, utils_1.getPluginConfigMetadata)(target);
        previousMetadata.set(metadata.name, metadata);
        Reflect.defineMetadata(constants_1.METADATA_KEY.config, previousMetadata, target.constructor);
    };
}
exports.config = config;
function playerdata(metadata) {
    if (metadata === void 0) { metadata = {}; }
    return function (target, key) {
        var _a, _b, _c, _d, _e;
        metadata.name = metadata.name || key;
        metadata.variable = key;
        metadata.version = (_a = metadata.version) !== null && _a !== void 0 ? _a : 1;
        metadata.format = (_b = metadata.format) !== null && _b !== void 0 ? _b : 'yml';
        metadata.autosave = (_c = metadata.autosave) !== null && _c !== void 0 ? _c : false;
        metadata.filename = (_d = metadata.filename) !== null && _d !== void 0 ? _d : "username";
        metadata.dir = (_e = metadata.dir) !== null && _e !== void 0 ? _e : "playerdata";
        var previousMetadata = (0, utils_1.getPluginPlayerDataMetadata)(target);
        previousMetadata.set(metadata.name, metadata);
        Reflect.defineMetadata(constants_1.METADATA_KEY.playerdata, previousMetadata, target.constructor);
    };
}
exports.playerdata = playerdata;
function stage(stage) {
    return function (metadata) {
        if (metadata === void 0) { metadata = {}; }
        return function (target, key, value) {
            metadata.name = metadata.name || key;
            metadata.executor = key;
            Reflect.defineMetadata(constants_1.METADATA_KEY.stage[stage], tslib_1.__spreadArray([metadata], tslib_1.__read((0, utils_1.getPluginStageMetadata)(target, stage)), false), target.constructor);
        };
    };
}
exports.load = stage('load');
exports.enable = stage('enable');
exports.disable = stage('disable');
