// @ccms/plugin/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlayerData = exports.Config = exports.Listener = exports.Tab = exports.Cmd = exports.JSPlugin = void 0;
var tslib_1 = require("tslib");
require("@ccms/verify");
require("./scanner/js-scanner");
require("./loader/ioc-loader");
require("./loader/basic-loader");
tslib_1.__exportStar(require("./config"), exports);
tslib_1.__exportStar(require("./config/config"), exports);
tslib_1.__exportStar(require("./manager"), exports);
tslib_1.__exportStar(require("./decorators"), exports);
tslib_1.__exportStar(require("./interfaces"), exports);
var decorators_1 = require("./decorators");
Object.defineProperty(exports, "JSPlugin", { enumerable: true, get: function () { return decorators_1.plugin; } });
Object.defineProperty(exports, "Cmd", { enumerable: true, get: function () { return decorators_1.cmd; } });
Object.defineProperty(exports, "Tab", { enumerable: true, get: function () { return decorators_1.tab; } });
Object.defineProperty(exports, "Listener", { enumerable: true, get: function () { return decorators_1.listener; } });
Object.defineProperty(exports, "Config", { enumerable: true, get: function () { return decorators_1.config; } });
Object.defineProperty(exports, "PlayerData", { enumerable: true, get: function () { return decorators_1.playerdata; } });
require("@ccms/database");
