// @ccms/plugin/dist/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.interfaces = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var utils_1 = require("./utils");
var File = Java.type('java.io.File');
var interfaces;
(function (interfaces) {
    var Plugin = /** @class */ (function () {
        function Plugin() {
            this.description = (0, utils_1.getPluginMetadata)(this);
        }
        Plugin.prototype.initialize = function () {
            // @ts-ignore
            this.logger = new this.Console(this.description.prefix || this.description.name);
        };
        Plugin.prototype.getDataFolder = function () {
            var parent = new File(this.description.source).parent;
            var dataFolder = new File(parent, this.description.name);
            return dataFolder.getAbsolutePath();
        };
        Plugin.prototype.load = function () { };
        Plugin.prototype.enable = function () { };
        Plugin.prototype.disable = function () { };
        tslib_1.__decorate([
            (0, container_1.inject)(api_1.server.Console),
            tslib_1.__metadata("design:type", api_1.MiaoScriptConsole)
        ], Plugin.prototype, "Console", void 0);
        tslib_1.__decorate([
            (0, container_1.postConstruct)(),
            tslib_1.__metadata("design:type", Function),
            tslib_1.__metadata("design:paramtypes", []),
            tslib_1.__metadata("design:returntype", void 0)
        ], Plugin.prototype, "initialize", null);
        Plugin = tslib_1.__decorate([
            (0, container_1.injectable)(),
            tslib_1.__metadata("design:paramtypes", [])
        ], Plugin);
        return Plugin;
    }());
    interfaces.Plugin = Plugin;
})(interfaces = exports.interfaces || (exports.interfaces = {}));
