// @ccms/plugin/dist/manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginManagerImpl = void 0;
var tslib_1 = require("tslib");
var i18n_1 = require("@ccms/i18n");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var interfaces_1 = require("./interfaces");
var task_1 = require("./task");
var event_1 = require("./event");
var config_1 = require("./config");
var command_1 = require("./command");
var Thread = Java.type('java.lang.Thread');
var PluginManagerImpl = /** @class */ (function () {
    function PluginManagerImpl() {
        this.initialized = false;
        this.sacnnerMap = new Map();
        this.loaderMap = new Map();
        this.instanceMap = new Map();
        this.metadataMap = new Map();
        this.lazyMetadataMap = new Map();
        // ignore unused
        this.taskManager;
        this.eventManager;
        this.configManager;
        this.commandManager;
    }
    PluginManagerImpl.prototype.initialize = function () {
        var _this = this;
        if (this.pluginInstance === undefined) {
            throw new Error("Can't found Plugin Instance!");
        }
        if (this.initialized !== true) {
            process.emit('plugin.manager.before.initialize', this);
            console.i18n('ms.plugin.initialize', { plugin: this.pluginInstance, loader: Thread.currentThread().contextClassLoader });
            try {
                console.i18n('ms.plugin.event.map', { count: this.eventManager.mapEventName(), type: this.serverType });
            }
            catch (error) {
                console.i18n('ms.plugin.event.map.error', { type: this.serverType, error: error });
            }
            var pluginScanner = this.container.getAll(api_1.plugin.PluginScanner);
            pluginScanner.forEach(function (scanner) {
                console.debug("loading plugin sacnner ".concat(scanner.type, "..."));
                _this.sacnnerMap.set(scanner.type, scanner);
            });
            var pluginLoaders = this.container.getAll(api_1.plugin.PluginLoader);
            pluginLoaders.forEach(function (loader) {
                console.debug("loading plugin loader ".concat(loader.type, "..."));
                _this.loaderMap.set(loader.type, loader);
            });
            this.initialized = true;
            process.emit('plugin.manager.after.initialize', this);
        }
    };
    PluginManagerImpl.prototype.scan = function (folder) {
        var e_1, _a, e_2, _b;
        if (!folder) {
            throw new Error('plugin scan folder can\'t be empty!');
        }
        this.initialize();
        process.emit('plugin.manager.before.scan', folder, this);
        try {
            for (var _c = tslib_1.__values(this.sacnnerMap), _d = _c.next(); !_d.done; _d = _c.next()) {
                var _e = tslib_1.__read(_d.value, 2), scanner = _e[1];
                try {
                    console.i18n('ms.plugin.manager.scan', { scanner: scanner.type, folder: folder });
                    var loadMetadatas = scanner.scan(folder);
                    console.i18n('ms.plugin.manager.scan.finish', { scanner: scanner.type, folder: folder, size: loadMetadatas.length });
                    try {
                        for (var loadMetadatas_1 = (e_2 = void 0, tslib_1.__values(loadMetadatas)), loadMetadatas_1_1 = loadMetadatas_1.next(); !loadMetadatas_1_1.done; loadMetadatas_1_1 = loadMetadatas_1.next()) {
                            var loadMetadata = loadMetadatas_1_1.value;
                            try {
                                this.loadAndRequirePlugin(loadMetadata);
                            }
                            catch (error) {
                                console.console("\u00A7c\u626B\u63CF\u5668 \u00A74".concat(scanner.type, " \u00A7c\u6587\u4EF6 \u00A74").concat(loadMetadata.file.toString().replace(root, ''), " \u00A7c\u7F16\u8BD1\u5931\u8D25 \u8BF7\u63D0\u4F9B\u4E0B\u5217\u9519\u8BEF\u7ED9\u5F00\u53D1\u8005"));
                                console.ex(error);
                            }
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (loadMetadatas_1_1 && !loadMetadatas_1_1.done && (_b = loadMetadatas_1.return)) _b.call(loadMetadatas_1);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                }
                catch (error) {
                    console.ex(error);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
            }
            finally { if (e_1) throw e_1.error; }
        }
        process.emit('plugin.manager.after.scan', folder, this);
    };
    PluginManagerImpl.prototype.build = function () {
        process.emit('plugin.manager.before.build', this);
        this.buildPlugins();
        process.emit('plugin.manager.after.build', this);
    };
    PluginManagerImpl.prototype.logStage = function (plugin, stage) {
        console.i18n("ms.plugin.manager.stage", { stage: stage, plugin: plugin.description.name, version: plugin.description.version, author: plugin.description.author });
    };
    PluginManagerImpl.prototype.runPluginStage = function (plugin, stage) {
        var _a, _b;
        if (!plugin) {
            throw new Error("can't run runPluginStage ".concat(stage, " because plugin is ").concat(plugin));
        }
        try {
            this.logStage(plugin, i18n_1.default.translate("ms.plugin.manager.stage.".concat(stage)));
            process.emit("plugin.before.".concat(stage), plugin);
            this.runCatch(plugin, stage);
            this.runCatch(plugin, "".concat(this.serverType).concat(stage));
            (_b = (_a = plugin.description.loadMetadata.loader)[stage]) === null || _b === void 0 ? void 0 : _b.call(_a, plugin);
            process.emit("plugin.after.".concat(stage), plugin);
        }
        catch (ex) {
            console.i18n("ms.plugin.manager.stage.exec.error", { plugin: plugin.description.name, executor: stage, error: ex });
            if (global.debug) {
                console.ex(ex);
            }
        }
    };
    PluginManagerImpl.prototype.loadPlugin = function (loadMetadata) {
        var e_3, _a;
        var _b, _c;
        if (!loadMetadata) {
            throw new Error('loadMetadata can\'t be undefiend when loadPlugin!');
        }
        if (loadMetadata.loaded) {
            throw new Error("Plugin file ".concat(loadMetadata.file, " is already loaded by ").concat((_b = loadMetadata.loader) === null || _b === void 0 ? void 0 : _b.type, "!"));
        }
        process.emit("plugin.before.require", loadMetadata);
        try {
            try {
                for (var _d = tslib_1.__values(this.loaderMap), _e = _d.next(); !_e.done; _e = _d.next()) {
                    var _f = tslib_1.__read(_e.value, 2), loader = _f[1];
                    if ((_c = this.loaderRequirePlugin(loadMetadata, loader)) === null || _c === void 0 ? void 0 : _c.loaded)
                        return loadMetadata.metadata;
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        catch (error) {
            console.i18n("ms.plugin.manager.initialize.error", { name: loadMetadata.file, ex: error });
            console.ex(error);
        }
    };
    PluginManagerImpl.prototype.loaderRequirePlugin = function (loadMetadata, loader) {
        try {
            if (loader.require(loadMetadata).loaded) {
                loadMetadata.loader = loader;
                var metadata = loadMetadata.metadata;
                if (this.metadataMap.has(metadata.name) && this.instanceMap.has(metadata.name)) {
                    var oldMetadata = this.metadataMap.get(metadata.name);
                    throw new Error("Plugin ".concat(oldMetadata.name, " is already load from ").concat(oldMetadata.source, "..."));
                }
                this.metadataMap.set(metadata.name, metadata);
                metadata.loadMetadata = loadMetadata;
                process.emit("plugin.after.require", loadMetadata);
            }
            return loadMetadata;
        }
        catch (error) {
            if (global.debug) {
                console.console("\u00A76Loader \u00A7b".concat(loader.type, " \u00A76load \u00A7a").concat(loadMetadata.file, " \u00A7cerror. \u00A74Err: \u00A7c").concat(error));
                console.ex(error);
            }
            else {
                console.warn("Loader ".concat(loader.type, " load ").concat(loadMetadata.file, " error. Err: ").concat(error));
            }
        }
    };
    PluginManagerImpl.prototype.loadAndRequirePlugin = function (loadMetadata) {
        var startTime = Date.now();
        var metadata = this.loadPlugin(loadMetadata.scanner.load(loadMetadata));
        if (!metadata) {
            throw new Error('load plugin metadata failed.');
        }
        console.i18n('ms.plugin.manager.build', {
            name: loadMetadata.metadata.name,
            version: loadMetadata.metadata.version,
            file: loadMetadata.file.toString().replace(root, ''),
            scanner: loadMetadata.scanner.type,
            loader: loadMetadata.loader.type,
            cost: (Date.now() - startTime) / 1000
        });
        return metadata;
    };
    /**
     * 从文件加载插件
     * @param file java.io.File
     */
    PluginManagerImpl.prototype.loadFromFile = function (file, ext) {
        if (ext === void 0) { ext = 'js'; }
        if (!file) {
            throw new Error('plugin file can\'t be undefiend.');
        }
        var scanner = this.sacnnerMap.get(ext);
        if (!scanner) {
            throw new Error("plugin scanner ".concat(ext, " can't found in sacnnerMap."));
        }
        var metadata = this.loadAndRequirePlugin(scanner.read(file));
        this.buildPlugin(metadata);
        var plugin = metadata.target;
        if (!plugin) {
            throw new Error("plugin scanner ".concat(ext, " can't found in sacnnerMap."));
        }
        this.load(plugin);
        this.enable(plugin);
        return plugin;
    };
    PluginManagerImpl.prototype.load = function () {
        var _this = this;
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.checkAndGet(args[0]).forEach(function (plugin) { return _this.runPluginStage(plugin, 'load'); });
    };
    PluginManagerImpl.prototype.enable = function () {
        var _this = this;
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.checkAndGet(args[0]).forEach(function (plugin) { return _this.runPluginStage(plugin, 'enable'); });
    };
    PluginManagerImpl.prototype.disable = function () {
        var _this = this;
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.checkAndGet(args[0]).forEach(function (plugin) {
            _this.runPluginStage(plugin, 'disable');
            _this.metadataMap.delete(plugin.description.name);
            _this.instanceMap.delete(plugin.description.name);
        });
    };
    PluginManagerImpl.prototype.reload = function () {
        var _this = this;
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.checkAndGet(args[0]).forEach(function (pl) {
            _this.disable(pl);
            _this.loadFromFile(pl.description.loadMetadata.file, pl.description.loadMetadata.scanner.type);
        });
    };
    PluginManagerImpl.prototype.has = function (name) {
        return this.instanceMap.has(name);
    };
    PluginManagerImpl.prototype.get = function (name) {
        return this.instanceMap.get(name) || null;
    };
    PluginManagerImpl.prototype.getPlugin = function (name) {
        return this.instanceMap.get(name) || null;
    };
    PluginManagerImpl.prototype.getPlugins = function () {
        return this.instanceMap;
    };
    PluginManagerImpl.prototype.runCatch = function (pl, func) {
        try {
            if (pl[func])
                pl[func].call(pl);
        }
        catch (ex) {
            console.i18n("ms.plugin.manager.stage.exec.error", { plugin: pl.description.name, executor: func, error: ex });
            console.ex(ex);
        }
    };
    PluginManagerImpl.prototype.checkAndGet = function (name) {
        var _a;
        if (name === undefined)
            throw new Error("checkAndGet Plugin can't be undefiend.");
        if (name == this.instanceMap) {
            return this.instanceMap;
        }
        if (typeof name == 'string' && this.instanceMap.has(name)) {
            return [this.instanceMap.get(name)];
        }
        if (name instanceof interfaces_1.interfaces.Plugin) {
            return [name];
        }
        if ((_a = name.description) === null || _a === void 0 ? void 0 : _a.name) {
            return [name];
        }
        throw new Error("Plugin ".concat(JSON.stringify(name), " not exist."));
    };
    PluginManagerImpl.prototype.buildPlugins = function () {
        var _this = this;
        this.metadataMap.forEach(function (metadata, key) {
            var _a;
            if ((_a = metadata === null || metadata === void 0 ? void 0 : metadata.depends) === null || _a === void 0 ? void 0 : _a.length) {
                _this.lazyMetadataMap.set(key, metadata);
            }
            else {
                _this.tryBuildPlugin(metadata);
            }
        });
        this.lazyMetadataMap.forEach(function (metadata, key) { return _this.tryBuildPlugin(metadata); });
    };
    PluginManagerImpl.prototype.checkDepends = function (depends) {
        var e_4, _a;
        if (!depends)
            return [];
        var loseDepends = [];
        try {
            for (var depends_1 = tslib_1.__values(depends), depends_1_1 = depends_1.next(); !depends_1_1.done; depends_1_1 = depends_1.next()) {
                var depend = depends_1_1.value;
                if (!this.metadataMap.has(depend))
                    loseDepends.push(depend);
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (depends_1_1 && !depends_1_1.done && (_a = depends_1.return)) _a.call(depends_1);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return loseDepends;
    };
    PluginManagerImpl.prototype.checkNativeDepends = function (depends) {
        var e_5, _a;
        if (!depends)
            return [];
        var loseDepends = [];
        try {
            for (var depends_2 = tslib_1.__values(depends), depends_2_1 = depends_2.next(); !depends_2_1.done; depends_2_1 = depends_2.next()) {
                var depend = depends_2_1.value;
                if (!this.nativePluginManager.has(depend))
                    loseDepends.push(depend);
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (depends_2_1 && !depends_2_1.done && (_a = depends_2.return)) _a.call(depends_2);
            }
            finally { if (e_5) throw e_5.error; }
        }
        return loseDepends;
    };
    PluginManagerImpl.prototype.tryBuildPlugin = function (metadata) {
        try {
            return this.buildPlugin(metadata);
        }
        catch (error) {
            console.ex(error);
        }
    };
    PluginManagerImpl.prototype.buildPlugin = function (metadata) {
        process.emit("plugin.before.build", metadata);
        if (this.instanceMap.has(metadata.name)) {
            throw new Error("Plugin ".concat(metadata.name, " is already load from ").concat(metadata.source, "..."));
        }
        if (!this.loaderMap.has(metadata.type)) {
            throw new Error("\u00A74\u65E0\u6CD5\u52A0\u8F7D\u63D2\u4EF6 \u00A7b".concat(metadata.name, " \u00A74\u8BF7\u68C0\u67E5 \u00A7c").concat(metadata.type, " \u00A74\u52A0\u8F7D\u5668\u662F\u5426\u6B63\u5E38\u542F\u7528!"));
        }
        if (!this.serverChecker.check(metadata.servers)) {
            throw new Error("\u00A76\u63D2\u4EF6 \u00A7b".concat(metadata.name, " \u00A7c\u670D\u52A1\u5668\u7C7B\u578B\u4E0D\u517C\u5BB9(").concat(metadata.servers.join(','), ") \u00A76\u5FFD\u7565\u52A0\u8F7D..."));
        }
        var loseDepends = this.checkDepends(metadata.depends) || [];
        if (loseDepends.length) {
            throw new Error("\u00A74\u65E0\u6CD5\u52A0\u8F7D\u63D2\u4EF6 \u00A7b".concat(metadata.name, " \u00A74\u8BF7\u68C0\u67E5\u811A\u672C\u4F9D\u8D56 \u00A73[").concat(loseDepends.join(','), "] \u00A74\u662F\u5426\u5B89\u88C5\u5B8C\u6574!"));
        }
        var loseNativeDepends = this.checkNativeDepends(metadata.nativeDepends) || [];
        if (loseNativeDepends.length) {
            throw new Error("\u00A74\u65E0\u6CD5\u52A0\u8F7D\u63D2\u4EF6 \u00A7b".concat(metadata.name, " \u00A74\u8BF7\u68C0\u67E5\u63D2\u4EF6\u4F9D\u8D56 \u00A73[").concat(loseNativeDepends.join(','), "] \u00A74\u662F\u5426\u5B89\u88C5\u5B8C\u6574!"));
        }
        var pluginInstance = this.loaderMap.get(metadata.type).build(metadata);
        if (!pluginInstance) {
            throw new Error("\u00A74\u52A0\u8F7D\u5668 \u00A7c".concat(metadata.type, " \u00A74\u52A0\u8F7D\u63D2\u4EF6 \u00A7c").concat(metadata.name, " \u00A74\u5931\u8D25!"));
        }
        metadata.target = pluginInstance;
        this.instanceMap.set(metadata.name, pluginInstance);
        process.emit("plugin.after.build", metadata, pluginInstance);
        return pluginInstance;
    };
    tslib_1.__decorate([
        (0, container_1.Autowired)(container_1.ContainerInstance),
        tslib_1.__metadata("design:type", container_1.Container)
    ], PluginManagerImpl.prototype, "container", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(api_1.plugin.PluginInstance),
        tslib_1.__metadata("design:type", Object)
    ], PluginManagerImpl.prototype, "pluginInstance", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(api_1.server.ServerType),
        tslib_1.__metadata("design:type", String)
    ], PluginManagerImpl.prototype, "serverType", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", api_1.server.ServerChecker)
    ], PluginManagerImpl.prototype, "serverChecker", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", api_1.server.NativePluginManager)
    ], PluginManagerImpl.prototype, "nativePluginManager", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", task_1.PluginTaskManager)
    ], PluginManagerImpl.prototype, "taskManager", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", event_1.PluginEventManager)
    ], PluginManagerImpl.prototype, "eventManager", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", config_1.PluginConfigManager)
    ], PluginManagerImpl.prototype, "configManager", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", command_1.PluginCommandManager)
    ], PluginManagerImpl.prototype, "commandManager", void 0);
    PluginManagerImpl = tslib_1.__decorate([
        (0, container_1.provideSingleton)(api_1.plugin.PluginManager),
        tslib_1.__metadata("design:paramtypes", [])
    ], PluginManagerImpl);
    return PluginManagerImpl;
}());
exports.PluginManagerImpl = PluginManagerImpl;
