// @ccms/plugin/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPluginStageMetadata = exports.getPluginPlayerDataMetadata = exports.getPluginConfigMetadata = exports.getPluginListenerMetadata = exports.getPluginTabCompleterMetadata = exports.getPluginCommandMetadata = exports.getPluginMetadata = exports.getPluginMetadatas = exports.getPluginSources = exports.getPlugins = exports.getPlugin = void 0;
var tslib_1 = require("tslib");
var constants_1 = require("./constants");
var pluginSourceCache = new Map();
function getPlugins() {
    return tslib_1.__spreadArray([], tslib_1.__read(getPluginMetadatas().values()), false).map(function (target) { return target.target; });
}
exports.getPlugins = getPlugins;
function getPlugin(name) {
    return getPluginMetadatas().get(name);
}
exports.getPlugin = getPlugin;
function getPluginSources() {
    var pluginSources = Reflect.getMetadata(constants_1.METADATA_KEY.source, Reflect) || pluginSourceCache;
    return pluginSources;
}
exports.getPluginSources = getPluginSources;
function getPluginMetadatas() {
    var pluginMetadatas = Reflect.getMetadata(constants_1.METADATA_KEY.plugin, Reflect) || new Map();
    return pluginMetadatas;
}
exports.getPluginMetadatas = getPluginMetadatas;
function getPluginMetadata(target) {
    var pluginMetadata = Reflect.getMetadata(constants_1.METADATA_KEY.plugin, target.constructor) || {};
    return pluginMetadata;
}
exports.getPluginMetadata = getPluginMetadata;
function getPluginCommandMetadata(target) {
    var commandMetadata = Reflect.getMetadata(constants_1.METADATA_KEY.cmd, target.constructor) || [];
    return commandMetadata;
}
exports.getPluginCommandMetadata = getPluginCommandMetadata;
function getPluginTabCompleterMetadata(target) {
    var tabcompleterMetadata = Reflect.getMetadata(constants_1.METADATA_KEY.tab, target.constructor) || new Map();
    return tabcompleterMetadata;
}
exports.getPluginTabCompleterMetadata = getPluginTabCompleterMetadata;
function getPluginListenerMetadata(target) {
    var listnerMetadata = Reflect.getMetadata(constants_1.METADATA_KEY.listener, target.constructor) || [];
    return listnerMetadata;
}
exports.getPluginListenerMetadata = getPluginListenerMetadata;
function getPluginConfigMetadata(target) {
    var configMetadata = Reflect.getMetadata(constants_1.METADATA_KEY.config, target.constructor) || new Map();
    return configMetadata;
}
exports.getPluginConfigMetadata = getPluginConfigMetadata;
function getPluginPlayerDataMetadata(target) {
    var playerdataMetadata = Reflect.getMetadata(constants_1.METADATA_KEY.playerdata, target.constructor) || new Map();
    return playerdataMetadata;
}
exports.getPluginPlayerDataMetadata = getPluginPlayerDataMetadata;
function getPluginStageMetadata(target, stage) {
    var stageMetadata = Reflect.getMetadata(constants_1.METADATA_KEY.stage[stage], target.constructor) || [];
    return stageMetadata;
}
exports.getPluginStageMetadata = getPluginStageMetadata;
