// @ccms/plugin/dist/loader/ioc-loader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IocLoader = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var interfaces_1 = require("../interfaces");
var utils_1 = require("../utils");
var LOADER_TYPE_NAME = 'ioc';
var IocLoader = /** @class */ (function () {
    function IocLoader() {
        this.type = LOADER_TYPE_NAME;
        this.pluginMetadataMap = (0, utils_1.getPluginSources)();
    }
    IocLoader.prototype.require = function (loadMetadata) {
        var metadata = this.pluginMetadataMap.get(loadMetadata.file.toString());
        if (metadata && metadata.type == this.type) {
            loadMetadata.metadata = metadata;
            loadMetadata.loaded = true;
        }
        return loadMetadata;
    };
    IocLoader.prototype.build = function (metadata) {
        var pluginInstance;
        try {
            this.bindPlugin(metadata);
            pluginInstance = this.container.getNamed(api_1.plugin.Plugin, metadata.name);
            if (!(pluginInstance instanceof interfaces_1.interfaces.Plugin)) {
                console.i18n('ms.plugin.manager.build.not.extends', { source: metadata.source });
                return;
            }
        }
        catch (ex) {
            console.i18n("ms.plugin.manager.initialize.error", { name: metadata.name, ex: ex });
            console.ex(ex);
        }
        return pluginInstance;
    };
    IocLoader.prototype.load = function (plugin) {
        this.stage(plugin, 'load');
    };
    IocLoader.prototype.enable = function (plugin) {
        this.stage(plugin, 'enable');
    };
    IocLoader.prototype.disable = function (plugin) {
        this.stage(plugin, 'disable');
    };
    IocLoader.prototype.reload = function (plugin) {
        this.disable(plugin);
        //@ts-ignore
        require(plugin.description.source, { cache: false });
        this.load(plugin);
        this.enable(plugin);
    };
    IocLoader.prototype.bindPlugin = function (metadata) {
        if (this.container.isBoundNamed(api_1.plugin.Plugin, metadata.name)) {
            var pluginInstance = this.container.getNamed(api_1.plugin.Plugin, metadata.name);
            if (pluginInstance.description.source + '' !== metadata.source + '') {
                console.i18n('ms.plugin.manager.build.duplicate', { exists: pluginInstance.description.source, source: metadata.source });
            }
            this.container.rebind(api_1.plugin.Plugin).to(metadata.target).inSingletonScope().whenTargetNamed(metadata.name);
        }
        else {
            this.container.bind(api_1.plugin.Plugin).to(metadata.target).inSingletonScope().whenTargetNamed(metadata.name);
        }
    };
    IocLoader.prototype.stage = function (pluginInstance, stageName) {
        var e_1, _a;
        var stages = (0, utils_1.getPluginStageMetadata)(pluginInstance, stageName);
        try {
            for (var stages_1 = tslib_1.__values(stages), stages_1_1 = stages_1.next(); !stages_1_1.done; stages_1_1 = stages_1.next()) {
                var stage = stages_1_1.value;
                if (!this.serverChecker.check(stage.servers)) {
                    continue;
                }
                console.i18n("ms.plugin.manager.stage.exec", { plugin: pluginInstance.description.name, name: stage.executor, stage: stageName, servers: stage.servers });
                try {
                    pluginInstance[stage.executor].apply(pluginInstance);
                }
                catch (error) {
                    console.i18n("ms.plugin.manager.stage.exec.error", { plugin: pluginInstance.description.name, executor: stage.executor, error: error });
                    console.ex(error);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (stages_1_1 && !stages_1_1.done && (_a = stages_1.return)) _a.call(stages_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    tslib_1.__decorate([
        (0, container_1.inject)(container_1.ContainerInstance),
        tslib_1.__metadata("design:type", container_1.Container)
    ], IocLoader.prototype, "container", void 0);
    tslib_1.__decorate([
        (0, container_1.Autowired)(),
        tslib_1.__metadata("design:type", api_1.server.ServerChecker)
    ], IocLoader.prototype, "serverChecker", void 0);
    IocLoader = tslib_1.__decorate([
        (0, container_1.provideSingletonNamed)(api_1.plugin.PluginLoader, LOADER_TYPE_NAME),
        tslib_1.__metadata("design:paramtypes", [])
    ], IocLoader);
    return IocLoader;
}());
exports.IocLoader = IocLoader;
