// @ccms/plugin/dist/config/file-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginConfigFolder = exports.PluginFileConfig = void 0;
var fs = require("@ccms/common/dist/fs");
var PluginFileConfig = /** @class */ (function () {
    function PluginFileConfig(loader, file, def) {
        if (def === void 0) { def = {}; }
        this.loader = loader;
        this.file = file;
        if (fs.exists(file)) {
            this.reload();
        }
        else {
            Object.assign(this, def);
        }
        this.initialize();
    }
    PluginFileConfig.prototype.initialize = function () {
    };
    PluginFileConfig.prototype.save = function () {
        base.save(this.file, this.loader.dump(this));
    };
    PluginFileConfig.prototype.reload = function () {
        Object.assign(this, this.loader.load(base.read(this.file)));
    };
    return PluginFileConfig;
}());
exports.PluginFileConfig = PluginFileConfig;
var PluginConfigFolder = /** @class */ (function () {
    function PluginConfigFolder(loader, folder) {
        this.configCache = new Map();
        this.loader = loader;
        this.folder = folder;
    }
    PluginConfigFolder.prototype.createConfig = function (path, def) {
        if (def === void 0) { def = {}; }
        return new PluginFileConfig(this.loader, path, def);
    };
    PluginConfigFolder.prototype.getConfig = function (name, def) {
        if (def === void 0) { def = {}; }
        var path = fs.concat(this.folder, name);
        if (!this.configCache.has(path)) {
            this.configCache.set(path, this.createConfig(path, def));
        }
        return this.configCache.get(path);
    };
    PluginConfigFolder.prototype.clear = function () {
        this.configCache.clear();
    };
    PluginConfigFolder.prototype.save = function () {
        this.configCache.forEach(function (config) { return config.save(); });
    };
    PluginConfigFolder.prototype.reload = function () {
        this.configCache.forEach(function (config) { return config.reload(); });
    };
    return PluginConfigFolder;
}());
exports.PluginConfigFolder = PluginConfigFolder;
