// @ccms/api/dist/channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.channel = void 0;
var tslib_1 = require("tslib");
var container_1 = require("@ccms/container");
var channel;
(function (channel_1) {
    var Channel = /** @class */ (function () {
        function Channel() {
            this.listenerMap = [];
        }
        /**
         * 注册通道
         * @param plugin 插件
         * @param channel 通道
         * @param exec 执行器
         */
        Channel.prototype.listen = function (plugin, channel, exec) {
            var _this = this;
            if (!plugin || !plugin.description || !plugin.description.name)
                throw new TypeError('Plugin can\'t be undefiend!');
            var name = plugin.description.name;
            var listener = this.register(channel, exec);
            if (!this.listenerMap[name])
                this.listenerMap[name] = [];
            var offExec = function () {
                _this.unregister(channel, listener);
                console.debug("[".concat(name, "] unregister channel ").concat(channel));
            };
            var off = {
                channel: channel,
                listener: listener,
                off: offExec
            };
            this.listenerMap[name].push(off);
            console.debug("[".concat(name, "] register channel ").concat(channel, " => ").concat(exec.name || '[anonymous]'));
            return off;
        };
        /**
         * 关闭插件注册的通道
         * @param plugin 插件
         */
        Channel.prototype.disable = function (plugin) {
            var channelCache = this.listenerMap[plugin.description.name];
            if (channelCache) {
                channelCache.forEach(function (t) { return t.off(); });
                delete this.listenerMap[plugin.description.name];
            }
        };
        Channel = tslib_1.__decorate([
            (0, container_1.injectable)()
        ], Channel);
        return Channel;
    }());
    channel_1.Channel = Channel;
})(channel = exports.channel || (exports.channel = {}));
