// @ccms/api/dist/chat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chat = void 0;
var tslib_1 = require("tslib");
var container_1 = require("@ccms/container");
var chat;
(function (chat) {
    var Chat = /** @class */ (function () {
        function Chat() {
        }
        /**
         * sendJsonChat
         * @param sender reciver
         * @param json json
         * @param type chat Type 0: chat 1: system 2: actionBar
         */
        Chat.prototype.sendJson = function (sender, json, type) {
            if (type === void 0) { type = 0; }
            throw new Error("Method not implemented.");
        };
        Chat.prototype.sendMessage = function (sender, message) {
            throw new Error("Method not implemented.");
        };
        Chat.prototype.sendActionBar = function (sender, message) {
            throw new Error("Method not implemented.");
        };
        Chat.prototype.clearActionBar = function (sender) {
            this.sendActionBar(sender, '');
        };
        Chat.prototype.sendTitle = function (sender, title, subtitle, fadeIn, time, fadeOut) {
            if (subtitle === void 0) { subtitle = ''; }
            if (fadeIn === void 0) { fadeIn = 20; }
            if (time === void 0) { time = 100; }
            if (fadeOut === void 0) { fadeOut = 20; }
            throw new Error("Method not implemented.");
        };
        Chat.prototype.clearTitle = function (sender) {
            this.sendTitle(sender, '', '', 0, 0, 0);
        };
        Chat = tslib_1.__decorate([
            (0, container_1.injectable)()
        ], Chat);
        return Chat;
    }());
    chat.Chat = Chat;
})(chat = exports.chat || (exports.chat = {}));
