// @ccms/api/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constants = void 0;
var constants;
(function (constants) {
    var ServiceIdentifier;
    (function (ServiceIdentifier) {
        /**
         * Runtime Server NettyPipeline
         */
        ServiceIdentifier.NettyPipeline = Symbol("NettyPipeline");
        /**
         * Runtime Server RootLogger
         */
        ServiceIdentifier.RootLogger = Symbol("RootLogger");
    })(ServiceIdentifier = constants.ServiceIdentifier || (constants.ServiceIdentifier = {}));
    var Reflect;
    (function (Reflect) {
        Reflect.Method = {
            getServerConnection: [
                /*spigot 1.8.8*/ 'aq',
                /*spigot 1.12.2*/ 'an',
                /*spigot 1.14.4+*/ 'getServerConnection',
                /*spigot 1.19+*/ 'ad',
                /*catserver 1.12.2*/ 'func_147137_ag'
            ]
        };
        Reflect.Field = {
            listeningChannels: [
                /*spigot 1.8.8-1.12.2*/ 'g',
                /*spigot 1.14.4*/ 'f',
                /*spigot 1.15.2+*/ 'listeningChannels',
                /*catserver 1.12.2*/ 'field_151274_e'
            ]
        };
    })(Reflect = constants.Reflect || (constants.Reflect = {}));
    var ServerType;
    (function (ServerType) {
        ServerType["Bukkit"] = "bukkit";
        ServerType["Sponge"] = "sponge";
        ServerType["Nukkit"] = "nukkit";
        ServerType["Bungee"] = "bungee";
        ServerType["Spring"] = "spring";
    })(ServerType = constants.ServerType || (constants.ServerType = {}));
})(constants = exports.constants || (exports.constants = {}));
