// @ccms/api/dist/particle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.particle = void 0;
var tslib_1 = require("tslib");
var container_1 = require("@ccms/container");
var index_1 = require("./index");
var UUID = Java.type('java.util.UUID');
var Math = Java.type('java.lang.Math');
var particle;
(function (particle_1) {
    /**
     * 表示一个特效对象
     *
     * @author Zoyn
     */
    var Particle = /** @class */ (function () {
        function Particle() {
            this.count = 1;
            this.offsetX = 0;
            this.offsetY = 0;
            this.offsetZ = 0;
            this.extra = 0;
            this.data = null;
            this.uuid = UUID.randomUUID().toString();
        }
        Particle.prototype.getUUID = function () {
            return this.uuid;
        };
        Particle.prototype.getSpawner = function () {
            return this.spawner;
        };
        Particle.prototype.setSpawner = function (spawner) {
            this.spawner = spawner;
            return this;
        };
        Particle.prototype.getParticle = function () {
            return this.particle;
        };
        Particle.prototype.setParticle = function (particle, data) {
            if (data === void 0) { data = null; }
            this.particle = particle;
            this.data = data;
            return this;
        };
        Particle.prototype.getCount = function () {
            return this.count;
        };
        Particle.prototype.setCount = function (count) {
            this.count = count;
            return this;
        };
        Particle.prototype.getOffsetX = function () {
            return this.offsetX;
        };
        Particle.prototype.setOffsetX = function (offsetX) {
            this.offsetX = offsetX;
            return this;
        };
        Particle.prototype.getOffsetY = function () {
            return this.offsetY;
        };
        Particle.prototype.setOffsetY = function (offsetY) {
            this.offsetY = offsetY;
            return this;
        };
        Particle.prototype.getOffsetZ = function () {
            return this.offsetZ;
        };
        Particle.prototype.setOffsetZ = function (offsetZ) {
            this.offsetZ = offsetZ;
            return this;
        };
        Particle.prototype.getExtra = function () {
            return this.extra;
        };
        Particle.prototype.setExtra = function (extra) {
            this.extra = extra;
            return this;
        };
        Particle.prototype.getData = function () {
            return this.data;
        };
        Particle.prototype.setData = function (data) {
            this.data = data;
            return this;
        };
        /**
         * 通过给定一个坐标就可以使用已经指定的参数来播放粒子
         *
         * @param location 坐标
         */
        Particle.prototype.spawn = function (location) {
            if (!this.spawner)
                throw new Error("particle ".concat(this.uuid, " not set spawner can't spawn!"));
            this.spawner.spawn(location, this);
        };
        return Particle;
    }());
    particle_1.Particle = Particle;
    /**
     * 表示一条线
     *
     * @author Zoyn
     */
    var Line = /** @class */ (function (_super) {
        tslib_1.__extends(Line, _super);
        /**
         * 构造一条线
         *
         * @param start  线的起点
         * @param end    线的终点
         * @param step   每个粒子之间的间隔 (也即步长)
         */
        function Line(start, end, step) {
            if (step === void 0) { step = 0.1; }
            var _this = _super.call(this) || this;
            _this.start = start;
            _this.end = end;
            _this.step = step;
            // 对向量进行重置
            _this.resetVector();
            return _this;
        }
        Line.prototype.show = function () {
            for (var i = 0; i < this.length; i += this.step) {
                var vectorTemp = this.vector.clone().multiply(i);
                this.spawn(this.start.clone().add(vectorTemp));
            }
        };
        /**
         * 获取线的起始坐标
         *
         * @return {@link Location}
         */
        Line.prototype.getStart = function () {
            return this.start;
        };
        /**
         * 利用给定的坐标设置线的起始坐标
         *
         * @param start 起始坐标
         * @return {@link Line}
         */
        Line.prototype.setStart = function (start) {
            this.start = start;
            this.resetVector();
            return this;
        };
        /**
         * 获取线的终点坐标
         *
         * @return {@link Location}
         */
        Line.prototype.getEnd = function () {
            return this.end;
        };
        /**
         * 利用给定的坐标设置线的终点坐标
         *
         * @param end 终点
         * @return {@link Line}
         */
        Line.prototype.setEnd = function (end) {
            this.end = end;
            this.resetVector();
            return this;
        };
        /**
         * 获取每个粒子之间的间隔
         *
         * @return 也就是循环的步长
         */
        Line.prototype.getStep = function () {
            return this.step;
        };
        /**
         * 设置每个粒子之间的间隔
         *
         * @param step 间隔
         * @return {@link Line}
         */
        Line.prototype.setStep = function (step) {
            this.step = step;
            this.resetVector();
            return this;
        };
        /**
         * 手动重设线的向量
         */
        Line.prototype.resetVector = function () {
            this.vector = this.end.clone().subtract(this.start).toVector();
            this.length = this.vector.length();
            this.vector.normalize();
        };
        Line.buildLine = function (locA, locB, step, particle) {
            var vectorAB = locB.clone().subtract(locA).toVector();
            var vectorLength = vectorAB.length();
            vectorAB.normalize();
            for (var i = 0; i < vectorLength; i += step) {
                ParticleManager.globalSpawner.spawn(locA.clone().add(vectorAB.clone().multiply(i)), particle);
            }
        };
        return Line;
    }(Particle));
    particle_1.Line = Line;
    /**
     * 表示一个弧
     *
     * @author Zoyn
     */
    var Arc = /** @class */ (function (_super) {
        tslib_1.__extends(Arc, _super);
        /**
         * 构造一个弧
         *
         * @param origin 弧所在的圆的圆点
         * @param angle  弧所占的角度
         * @param radius 弧所在的圆的半径
         * @param step   每个粒子的间隔(也即步长)
         */
        function Arc(angle, radius, step) {
            if (angle === void 0) { angle = 30; }
            if (radius === void 0) { radius = 1; }
            if (step === void 0) { step = 1; }
            var _this = _super.call(this) || this;
            _this.angle = angle;
            _this.radius = radius;
            _this.step = step;
            return _this;
        }
        Arc.prototype.show = function (location) {
            for (var i = 0; i < this.angle; i += this.step) {
                var radians = Math.toRadians(i);
                var x = this.radius * Math.cos(radians);
                var z = this.radius * Math.sin(radians);
                _super.prototype.spawn.call(this, location.clone().add(x, 0, z));
            }
        };
        Arc.prototype.getAngle = function () {
            return this.angle;
        };
        Arc.prototype.setAngle = function (angle) {
            this.angle = angle;
            return this;
        };
        Arc.prototype.getRadius = function () {
            return this.radius;
        };
        Arc.prototype.setRadius = function (radius) {
            this.radius = radius;
            return this;
        };
        Arc.prototype.getStep = function () {
            return this.step;
        };
        Arc.prototype.setStep = function (step) {
            this.step = step;
            return this;
        };
        return Arc;
    }(Particle));
    particle_1.Arc = Arc;
    /**
     * 表示一个圆
     *
     * @author Zoyn
     */
    var Circle = /** @class */ (function (_super) {
        tslib_1.__extends(Circle, _super);
        /**
         * 构造一个圆
         *
         * @param origin 圆的圆点
         * @param radius 圆的半径
         * @param step   每个粒子的间隔(也即步长)
         * @param period 特效周期(如果需要可以使用)
         */
        function Circle(radius, step) {
            if (radius === void 0) { radius = 1; }
            if (step === void 0) { step = 1; }
            // Circle只需要控制这个fullArc就可以满足所有的要求
            return _super.call(this, 360, radius, step) || this;
        }
        return Circle;
    }(Arc));
    particle_1.Circle = Circle;
    var AtomicInteger = Java.type("java.util.concurrent.atomic.AtomicInteger");
    var ParticleManager = /** @class */ (function () {
        function ParticleManager() {
            this.cacheTasks = new Map();
            this.pluginCacheTasks = new Map();
            this.taskId = new AtomicInteger(0);
        }
        /**
         * 获得自增的任务ID
         */
        ParticleManager.prototype.nextId = function () {
            return this.taskId.incrementAndGet();
        };
        ParticleManager.prototype.getTaskManager = function () {
            return this.taskManager;
        };
        ParticleManager.prototype.create = function (particle, plugin) {
            var uuid = particle.getUUID();
            if (this.cacheTasks.has(uuid)) {
                return this.cacheTasks.get(uuid);
            }
            var task = this.create0(plugin, particle);
            this.cacheTasks.set(uuid, task);
            if (plugin) {
                if (!this.pluginCacheTasks.has(plugin.description.name)) {
                    this.pluginCacheTasks.set(plugin.description.name, new Map());
                }
                this.pluginCacheTasks.get(plugin.description.name).set(task.getTaskId(), task);
            }
            return task;
        };
        ParticleManager.prototype.cancel = function (particle) {
            var uuid = particle.getUUID();
            if (this.cacheTasks.has(uuid)) {
                this.cacheTasks.get(uuid).cancel();
                this.cacheTasks.delete(uuid);
            }
            else {
                throw new Error("particle ".concat(uuid, " not found!"));
            }
        };
        ParticleManager.prototype.disable = function (plugin) {
            if (this.pluginCacheTasks.has(plugin.description.name)) {
                this.pluginCacheTasks
                    .get(plugin.description.name)
                    .forEach(function (task) { return task.cancel(); });
                this.pluginCacheTasks.delete(plugin.description.name);
            }
        };
        ParticleManager.prototype.create0 = function (owner, particle) {
            particle.setSpawner(this.getGlobalSpawner());
            return new ParticleTask(owner, particle, this);
        };
        ParticleManager.globalSpawner = undefined;
        tslib_1.__decorate([
            (0, container_1.Autowired)(),
            tslib_1.__metadata("design:type", index_1.task.TaskManager)
        ], ParticleManager.prototype, "taskManager", void 0);
        ParticleManager = tslib_1.__decorate([
            (0, container_1.injectable)(),
            tslib_1.__metadata("design:paramtypes", [])
        ], ParticleManager);
        return ParticleManager;
    }());
    particle_1.ParticleManager = ParticleManager;
    var ParticleTask = /** @class */ (function () {
        function ParticleTask(owner, particle, particleManager) {
            this.isAsync = false;
            this.interval = 0;
            this.owner = owner;
            this.taskId = particleManager.nextId();
            this.particle = particle;
            this.taskManager = particleManager.getTaskManager();
            this.particleManager = particleManager;
        }
        ParticleTask.prototype.getOwner = function () {
            return this.task.getOwner();
        };
        ParticleTask.prototype.getTaskId = function () {
            return this.taskId;
        };
        ParticleTask.prototype.getParticle = function () {
            return this.particle;
        };
        ParticleTask.prototype.async = function (isAsync) {
            if (isAsync === void 0) { isAsync = true; }
            this.isAsync = isAsync;
            return this;
        };
        ParticleTask.prototype.timer = function (tick) {
            this.interval = tick;
            return this;
        };
        ParticleTask.prototype.follow = function (entity) {
            this._follow = entity;
            this._location = undefined;
            return this;
        };
        ParticleTask.prototype.location = function (location) {
            this._location = location;
            this._follow = undefined;
            return this;
        };
        ParticleTask.prototype.submit = function () {
            var _this = this;
            this.cancel();
            if (this._follow && !this.interval)
                throw new Error("enable follow entity but interval is ".concat(this.interval, "!"));
            this.taskManager.create(function () {
                _this.task = _this.taskManager
                    .create(function () {
                    try {
                        if (_this._follow) {
                            if (!_this._follow.isOnline())
                                return _this.cancel();
                            _this.particle.show(_this._follow.getLocation().clone().add(0, 1, 0));
                        }
                        else if (_this._location) {
                            _this.particle.show(_this._location);
                        }
                        else {
                            console.warn("ParticleTask ".concat(_this.taskId, " particle ").concat(_this.particle.getUUID(), " cancel because entity and location both undefined!"));
                            _this.task.cancel();
                        }
                    }
                    catch (error) {
                        console.error("\u00A76\u63D2\u4EF6 \u00A7a".concat(_this.owner.description.name, " \u00A7c\u64AD\u653E\u7C92\u5B50\u53D1\u9001\u5F02\u5E38 \u00A74\u7C92\u5B50\u64AD\u653E\u4EFB\u52A1\u5DF2\u7EC8\u6B62!"));
                        console.ex(error);
                        _this.cancel();
                    }
                }, _this.owner)
                    .async(_this.isAsync)
                    .timer(_this.interval)
                    .submit();
            }, this.owner).later(2).submit();
            return this;
        };
        ParticleTask.prototype.cancel = function () {
            if (this.task != null) {
                this.task.cancel();
            }
        };
        return ParticleTask;
    }());
    particle_1.ParticleTask = ParticleTask;
    var ParticleSpawner = /** @class */ (function () {
        function ParticleSpawner() {
        }
        return ParticleSpawner;
    }());
    particle_1.ParticleSpawner = ParticleSpawner;
})(particle = exports.particle || (exports.particle = {}));
