// @ccms/api/dist/task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.task = void 0;
var tslib_1 = require("tslib");
var container_1 = require("@ccms/container");
var AtomicInteger = Java.type("java.util.concurrent.atomic.AtomicInteger");
var task;
(function (task_1) {
    var TaskManager = /** @class */ (function () {
        function TaskManager() {
            var _this = this;
            this.cacheTasks = new Map();
            this.pluginCacheTasks = new Map();
            this.taskId = new AtomicInteger(0);
            process.on('task.finish', function (task) {
                var _a, _b;
                var taskId = task.getTaskId();
                _this.cacheTasks.delete(taskId);
                var ownerName = (_a = task.getOwner()) === null || _a === void 0 ? void 0 : _a.description.name;
                if (ownerName && _this.pluginCacheTasks.has(ownerName)) {
                    (_b = _this.pluginCacheTasks.get(ownerName)) === null || _b === void 0 ? void 0 : _b.delete(taskId);
                }
            });
        }
        TaskManager.prototype.pluginCreate = function (plugin, task) {
            if (!this.pluginCacheTasks.has(plugin.description.name)) {
                this.pluginCacheTasks.set(plugin.description.name, new Map());
            }
            this.pluginCacheTasks.get(plugin.description.name).set(task.getTaskId(), task);
            return task;
        };
        TaskManager.prototype.pluginDisable = function (plugin) {
            if (this.pluginCacheTasks.has(plugin.description.name)) {
                this.pluginCacheTasks.get(plugin.description.name).forEach(function (task) {
                    task.cancel();
                });
                this.pluginCacheTasks.delete(plugin.description.name);
            }
        };
        /**
         * 获得自增的任务ID
         */
        TaskManager.prototype.nextId = function () {
            return this.taskId.incrementAndGet();
        };
        /**
         * 创建任务
         * @param func 任务内容
         */
        TaskManager.prototype.create = function (func, plugin) {
            if (Object.prototype.toString.call(func) !== "[object Function]") {
                throw TypeError('第一个参数 Task 必须为 function !');
            }
            ;
            var task = this.create0(plugin, func, this.nextId());
            this.cacheTasks.set(task.getTaskId(), task);
            if (plugin) {
                return this.pluginCreate(plugin, task);
            }
            return task;
        };
        TaskManager.prototype.cancel = function (taskId) {
            if (!this.cacheTasks.has(taskId)) {
                throw new Error("Task ".concat(taskId, " not found!"));
            }
            this.cacheTasks.get(taskId).cancel();
        };
        /**
         * 关闭任务管理器
         */
        TaskManager.prototype.disable = function (plugin) {
            if (plugin) {
                return this.pluginDisable(plugin);
            }
            this.disable0();
        };
        TaskManager = tslib_1.__decorate([
            (0, container_1.injectable)(),
            tslib_1.__metadata("design:paramtypes", [])
        ], TaskManager);
        return TaskManager;
    }());
    task_1.TaskManager = TaskManager;
    /**
     * 任务抽象
     */
    var Task = /** @class */ (function () {
        function Task(owner, func, id) {
            this.isAsync = false;
            this.laterTime = 0;
            this.interval = 0;
            this.owner = owner;
            this.func = func;
            this.taskId = id;
        }
        Task.prototype.getOwner = function () {
            return this.owner;
        };
        Task.prototype.getTaskId = function () {
            return this.taskId;
        };
        /**
         * 设置任务异步执行
         * @param isAsync 是否异步
         */
        Task.prototype.async = function (isAsync) {
            if (isAsync === void 0) { isAsync = true; }
            this.isAsync = isAsync;
            return this;
        };
        /**
         * 设置任务延时执行
         * @param tick 延时 Tick
         */
        Task.prototype.later = function (tick) {
            this.laterTime = tick;
            return this;
        };
        /**
         * 设置任务循环执行
         * @param tick 循环时间 Tick
         */
        Task.prototype.timer = function (tick) {
            this.interval = tick;
            return this;
        };
        /**
         * 取消任务
         */
        Task.prototype.cancel = function () {
            var result = this.cancel0();
            process.emit('task.finish', this);
            return result;
        };
        Task.prototype.run = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            try {
                this.func.apply(this, tslib_1.__spreadArray([], tslib_1.__read(args), false));
                !this.interval && process.emit('task.finish', this);
            }
            catch (ex) {
                console.console('§4插件执行任务时发生错误', ex);
                console.ex(ex);
            }
        };
        /**
         * 提交任务
         * @param args 任务参数
         */
        Task.prototype.submit = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            this.innerTask = this.submit0.apply(this, tslib_1.__spreadArray([], tslib_1.__read(args), false));
            return this;
        };
        /**
         * 取消任务
         */
        Task.prototype.cancel0 = function () {
            var _a;
            return (_a = this.innerTask) === null || _a === void 0 ? void 0 : _a.cancel();
        };
        return Task;
    }());
    task_1.Task = Task;
})(task = exports.task || (exports.task = {}));
