// @ccms/plugin/dist/scanner/js-scanner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSFileScanner = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var fs = require("@ccms/common/dist/fs");
var container_1 = require("@ccms/container");
var SCANNER_TYPE_NAME = 'js';
var JSFileScanner = /** @class */ (function () {
    function JSFileScanner() {
        this.type = SCANNER_TYPE_NAME;
    }
    JSFileScanner.prototype.scan = function (target) {
        var _this = this;
        return this.scanFolder(fs.concat(root, target)).map(function (file) { return _this.read(file); });
    };
    JSFileScanner.prototype.read = function (file) {
        return { file: file, type: this.type, scanner: this, loaded: false };
    };
    JSFileScanner.prototype.load = function (metadata) {
        if (metadata.type !== this.type) {
            return;
        }
        this.updatePlugin(metadata.file);
        //@ts-ignore load plugin not use cache
        metadata.instance = require(metadata.file.toString(), { cache: false });
        return metadata;
    };
    JSFileScanner.prototype.scanFolder = function (folder) {
        var files = [];
        this.checkUpdateFolder(folder);
        // must check file is exist maybe is a illegal symbolic link file
        fs.list(folder).forEach(function (path) {
            var file = path.toFile();
            if (file.exists() && file.getName().endsWith(".js")) {
                files.push(file);
            }
        });
        return files;
    };
    JSFileScanner.prototype.checkUpdateFolder = function (path) {
        var update = fs.file(path, "update");
        if (!update.exists()) {
            update.mkdirs();
        }
    };
    JSFileScanner.prototype.updatePlugin = function (file) {
        var update = fs.file(fs.file(fs.file(file).parentFile, 'update'), file.name);
        if (update.exists()) {
            console.i18n("ms.plugin.manager.build.update", { name: file.name });
            fs.move(update, file, true);
        }
    };
    JSFileScanner = tslib_1.__decorate([
        (0, container_1.provideSingletonNamed)(api_1.plugin.PluginScanner, SCANNER_TYPE_NAME)
    ], JSFileScanner);
    return JSFileScanner;
}());
exports.JSFileScanner = JSFileScanner;
