// @ccms/polyfill/dist/node-shim.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var events_1 = require("events");
var System = Java.type('java.lang.System');
var Thread = Java.type('java.lang.Thread');
var ManagementFactory = Java.type('java.lang.management.ManagementFactory');
var InterruptedException = Java.type('java.lang.InterruptedException');
var ThreadGroup = Java.type("java.lang.ThreadGroup");
var AtomicInteger = Java.type("java.util.concurrent.atomic.AtomicInteger");
var Callable = Java.type('java.util.concurrent.Callable');
var ThreadFactory = Java.type("java.util.concurrent.ThreadFactory");
var TimeoutException = Java.type('java.util.concurrent.TimeoutException');
var ThreadPoolExecutor = Java.type('java.util.concurrent.ThreadPoolExecutor');
var LinkedBlockingQueue = Java.type("java.util.concurrent.LinkedBlockingQueue");
var TimeUnit = Java.type('java.util.concurrent.TimeUnit');
var DelayQueue = Java.type('java.util.concurrent.DelayQueue');
var JavaScriptTask = Java.type(base.getJavaScriptTaskClass().name);
var threadCount = new AtomicInteger(0);
var threadGroup = new ThreadGroup("@ccms/micro-task");
var microTaskPool = new ThreadPoolExecutor(100, 200, 60, TimeUnit.SECONDS, new LinkedBlockingQueue(1024), new ThreadFactory(function (run) { return new Thread(threadGroup, run, "@ccms/micro-task-" + threadCount.incrementAndGet()); }));
var Process = /** @class */ (function (_super) {
    tslib_1.__extends(Process, _super);
    function Process() {
        var _this = _super.call(this) || this;
        _this.version = base.version;
        _this.versions = [];
        _this.config = {};
        _this.pid = parseInt(ManagementFactory.getRuntimeMXBean().getName().split('@')[0]);
        _this.arch = System.getProperty("os.arch");
        _this.platform = System.getProperty("os.name");
        _this.env = {
            __noSuchProperty__: function (prop) {
                return System.getenv(prop);
            }
        };
        _this.stdout = System.out;
        _this.stderr = System.err;
        _this.stdin = System.in;
        _this.execArgv = '';
        _this.execPath = '';
        _this.exitCode = 0;
        _this.on('exit', function () {
            console.log("await microTaskPool termination! queueTask: ".concat(microTaskPool.shutdownNow().size(), " remainTask: ").concat(threadGroup.activeCount()));
            microTaskPool.awaitTermination(3000, TimeUnit.MILLISECONDS);
        });
        return _this;
    }
    Process.prototype.on = function (event, listener) {
        var _this = this;
        return _super.prototype.on.call(this, event, function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            try {
                listener.apply(void 0, tslib_1.__spreadArray([], tslib_1.__read(args), false));
            }
            catch (origin) {
                try {
                    _super.prototype.emit.call(_this, 'error', origin);
                }
                catch (error) {
                    console.ex(origin);
                    console.ex(error);
                }
            }
        });
    };
    Process.prototype.nextTick = function (callback) {
        var _this = this;
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        microTaskPool.execute(function () {
            try {
                callback(args);
            }
            catch (origin) {
                try {
                    _super.prototype.emit.call(_this, 'error', origin);
                }
                catch (error) {
                    console.ex(origin);
                    console.ex(error);
                }
            }
        });
    };
    Process.prototype.exit = function (code) {
        console.log("process exit by code ".concat(code, "!"));
        this.emit('exit', this.exitCode = code);
    };
    Process.prototype.openStdin = function () {
        throw new Error('MiaoScript unsupport openStdin.');
    };
    Process.prototype.chdir = function (directory) {
        console.error('MiaoScript unsupport chdir. lock at ' + root);
    };
    Process.prototype.cwd = function () {
        return root;
    };
    Process.prototype.getgid = function () {
        throw new Error('MiaoScript unsupport getgid.');
    };
    Process.prototype.setgid = function (id) {
        throw new Error('MiaoScript unsupport setgid.');
    };
    Process.prototype.getuid = function () {
        throw new Error('MiaoScript unsupport getuid.');
    };
    Process.prototype.setuid = function (id) {
        throw new Error('MiaoScript unsupport setuid.');
    };
    Process.prototype.setUncaughtExceptionCaptureCallback = function (cb) {
        if (cb == null) {
            this.removeAllListeners('error');
        }
        else {
            this.on('error', cb);
        }
    };
    Process.prototype.hasUncaughtExceptionCaptureCallback = function () {
        return this.listenerCount('error') > 0;
    };
    Process.prototype.kill = function (pid, signal) {
        throw new Error('MiaoScript unsupport kill.');
        return true;
    };
    Process.prototype.toString = function () {
        return "[object process]";
    };
    return Process;
}(events_1.EventEmitter));
var EventLoop = /** @class */ (function () {
    function EventLoop() {
        var _this = this;
        this.eventLoopMainThread = undefined;
        this.eventLoopTaskQueue = new DelayQueue();
        this.taskExecuteTimeout = 3000;
        this.fixedThreadPool = undefined;
        this.timeoutCount = new AtomicInteger(0);
        this.timeoutTasks = [];
        this.intervalCount = new AtomicInteger(0);
        this.intervalTasks = [];
        this.taskExecuteTimeout = parseInt(process.env.MS_TASK_EXECUTE_TIMEOUT) || 3000;
        this.fixedThreadPool = new ThreadPoolExecutor(1, 1, 0, TimeUnit.SECONDS, new LinkedBlockingQueue(1024), new ThreadFactory(function (run) {
            var thread = new Thread(run, "@ccms/event-loop");
            thread.setDaemon(true);
            return thread;
        }));
        this.eventLoopMainThread = new Thread(function () {
            var task = undefined;
            try {
                while (!_this.eventLoopMainThread.isInterrupted()) {
                    task = _this.eventLoopTaskQueue.take();
                    try {
                        task.getTask()();
                    }
                    catch (cause) {
                        try {
                            process.emit('error', cause);
                        }
                        catch (error) {
                            console.error(cause);
                            console.ex(cause);
                        }
                    }
                }
            }
            catch (error) {
                console.log("EventLoop Thread isInterrupted exit! remainTask: ".concat(_this.eventLoopTaskQueue.size()));
                _this.eventLoopTaskQueue.clear();
                _this.eventLoopTaskQueue = undefined;
                _this.timeoutCount = undefined;
                _this.timeoutTasks = undefined;
                _this.intervalCount = undefined;
                _this.intervalTasks = undefined;
                _this.eventLoopMainThread = undefined;
            }
        }, "@ccms/event-loop");
        this.eventLoopMainThread.setDaemon(true);
        process.on('exit', function () {
            _this.eventLoopMainThread.interrupt();
            _this.fixedThreadPool.shutdownNow();
            console.log("await fixedThreadPool termination!");
            _this.fixedThreadPool.awaitTermination(3000, TimeUnit.MILLISECONDS);
        });
    }
    EventLoop.prototype.startEventLoop = function () {
        this.eventLoopMainThread.start();
    };
    EventLoop.prototype.putDelayTask = function (id, callback, ms) {
        this.eventLoopTaskQueue.put(new JavaScriptTask(id, callback, ms));
    };
    EventLoop.prototype.limitTimeTask = function (name, callback) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        if (!callback) {
            throw new Error("task ".concat(name, " callback function can't be null!"));
        }
        if (this.fixedThreadPool.isShutdown()) {
            return console.warn("FixedThreadPool isTerminated! ignore Task ".concat(name, "!"));
        }
        try {
            this.fixedThreadPool.submit(new Callable({
                call: function () {
                    try {
                        callback.apply(undefined, args);
                    }
                    catch (cause) {
                        cause = cause.getCause && cause.getCause() || cause;
                        try {
                            process.emit('error', cause);
                        }
                        catch (error) {
                            console.error(cause);
                            console.ex(cause);
                        }
                    }
                }
            })).get(this.taskExecuteTimeout, TimeUnit.MILLISECONDS);
        }
        catch (error) {
            if (error instanceof InterruptedException) {
                return console.warn("FixedThreadPool isInterrupted exit! Task ".concat(name, " exec exit!"));
            }
            if (error instanceof TimeoutException) {
                return console.warn("Task ".concat(name, " => ").concat(callback, " exec time greater than ").concat(this.taskExecuteTimeout, "s!"));
            }
            throw error.getCause && error.getCause() || error;
        }
    };
    EventLoop.prototype.setTimeout = function (callback, ms) {
        var _this = this;
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        var taskId = this.timeoutCount.incrementAndGet();
        this.timeoutTasks[taskId] = callback;
        console.trace("create setTimeout task ".concat(taskId, " => ").concat(callback));
        this.putDelayTask(taskId, function () {
            if (_this.timeoutTasks[taskId]) {
                _this.limitTimeTask.apply(_this, tslib_1.__spreadArray(["setTimeout-".concat(taskId), callback], tslib_1.__read(args), false));
            }
            else {
                console.trace("ignore setTimeout ".concat(ms, "ms task ").concat(taskId, " because it's cancelled!"));
            }
        }, ms);
        return taskId;
    };
    EventLoop.prototype.clearTimeout = function (taskId) {
        delete this.timeoutTasks[taskId];
    };
    EventLoop.prototype.setInterval = function (callback, ms) {
        var _this = this;
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        var taskId = this.intervalCount.incrementAndGet();
        this.intervalTasks[taskId] = callback;
        console.trace("create setInterval ".concat(ms, "ms task ").concat(taskId, " => ").concat(callback));
        var intervalTask = function () {
            if (_this.intervalTasks[taskId]) {
                _this.limitTimeTask.apply(_this, tslib_1.__spreadArray(["setInterval-".concat(taskId), callback], tslib_1.__read(args), false));
                _this.putDelayTask(taskId, intervalTask, ms);
            }
            else {
                console.trace("ignore setInterval task ".concat(taskId, " because it's cancelled!"));
            }
        };
        this.putDelayTask(taskId, intervalTask, ms);
        return taskId;
    };
    EventLoop.prototype.clearInterval = function (taskId) {
        delete this.intervalTasks[taskId];
    };
    return EventLoop;
}());
global.setGlobal('process', new Process(), {});
Object.defineProperty(process, require('core-js/es/symbol/to-string-tag'), { value: '[object process]' });
var eventLoop = new EventLoop();
Object.defineProperty(process, 'eventLoop', { value: eventLoop });
eventLoop.startEventLoop();
global.setGlobal('queueMicrotask', function (func) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    return process.nextTick(func, args);
}, {});
global.setGlobal('setTimeout', eventLoop.setTimeout.bind(eventLoop), {});
global.setGlobal('clearTimeout', eventLoop.clearTimeout.bind(eventLoop), {});
global.setGlobal('setInterval', eventLoop.setInterval.bind(eventLoop), {});
global.setGlobal('clearInterval', eventLoop.clearInterval.bind(eventLoop), {});
global.setGlobal('setImmediate', function (callback) {
    var args = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        args[_i - 1] = arguments[_i];
    }
    return eventLoop.setTimeout.apply(eventLoop, tslib_1.__spreadArray([callback, 0], tslib_1.__read(args), false));
}, { writable: true });
global.setGlobal('clearImmediate ', eventLoop.clearTimeout.bind(eventLoop), { writable: true });
