// @ccms/polyfill/dist/xml-http-request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.XMLHttpRequest = void 0;
require("@ccms/nashorn");
var URL = Java.type("java.net.URL");
var JavaString = Java.type("java.lang.String");
var SecureRandom = Java.type("java.security.SecureRandom");
var SSLContext = Java.type("javax.net.ssl.SSLContext");
var HttpsURLConnection = Java.type("javax.net.ssl.HttpsURLConnection");
var HostnameVerifier = Java.type("javax.net.ssl.HostnameVerifier");
var X509TrustManager = Java.type("javax.net.ssl.X509TrustManager");
var SocketTimeoutException = Java.type('java.net.SocketTimeoutException');
var Callable = Java.type('java.util.concurrent.Callable');
var TimeUnit = Java.type('java.util.concurrent.TimeUnit');
var Executors = Java.type('java.util.concurrent.Executors');
var ByteArrayOutputStream = Java.type("java.io.ByteArrayOutputStream");
var ByteArray = Java.type("byte[]");
var UTF_8 = "UTF-8";
var TrustAnyHostnameVerifier = new HostnameVerifier({ verify: function () { return true; } });
var SSLSocketFactory = function initSSLSocketFactory() {
    var sslContext = SSLContext.getInstance("TLS");
    sslContext.init(null, [new X509TrustManager({
            getAcceptedIssuers: function () { return null; },
            checkClientTrusted: function () { },
            checkServerTrusted: function () { }
        })], new SecureRandom());
    return sslContext.getSocketFactory();
}();
var ReadyState;
(function (ReadyState) {
    ReadyState[ReadyState["UNSENT"] = 0] = "UNSENT";
    ReadyState[ReadyState["OPENED"] = 1] = "OPENED";
    ReadyState[ReadyState["HEADERS_RECEIVED"] = 2] = "HEADERS_RECEIVED";
    ReadyState[ReadyState["LOADING"] = 3] = "LOADING";
    ReadyState[ReadyState["DONE"] = 4] = "DONE";
})(ReadyState || (ReadyState = {}));
var executor = Executors.newCachedThreadPool();
var XMLHttpRequest = /** @class */ (function () {
    function XMLHttpRequest() {
        this._connectTimeout = 5000;
        this._readTimeout = 120000;
        this._responseType = 'text';
        this._readyState = ReadyState.UNSENT;
        this._requestHeaders = {};
        this._status = 0;
        this._statusText = null;
        this._responseHeaders = {};
        this._connection = null;
    }
    Object.defineProperty(XMLHttpRequest.prototype, "timeout", {
        get: function () {
            return this._readTimeout;
        },
        set: function (timeout) {
            this._readTimeout = timeout;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "connectTimeout", {
        get: function () {
            return this._connectTimeout;
        },
        set: function (timeout) {
            this._connectTimeout = timeout;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "readTimeout", {
        get: function () {
            return this._readTimeout;
        },
        set: function (timeout) {
            this._readTimeout = timeout;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "readyState", {
        get: function () {
            return this._readyState;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "responseType", {
        get: function () {
            return this._responseType;
        },
        set: function (type) {
            this._responseType = type;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "status", {
        get: function () {
            return this._status;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "statusText", {
        get: function () {
            return this._statusText;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "response", {
        get: function () {
            return this._response || this.get();
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "responseText", {
        get: function () {
            return this._responseText;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "responseXML", {
        get: function () {
            return this._response;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(XMLHttpRequest.prototype, "responseURL", {
        get: function () {
            return this._responseURL;
        },
        enumerable: false,
        configurable: true
    });
    XMLHttpRequest.prototype.setRequestHeader = function (key, val) {
        this._requestHeaders[key] = val;
    };
    XMLHttpRequest.prototype.getResponseHeader = function (key) {
        return this._responseHeaders[key];
    };
    XMLHttpRequest.prototype.getAllResponseHeaders = function () {
        return this._responseHeaders;
    };
    XMLHttpRequest.prototype.addEventListener = function (event, callback) {
        this["on".concat(event.toLowerCase())] = callback;
    };
    XMLHttpRequest.prototype.overrideMimeType = function (mimeType) {
        this._mimeType = mimeType;
    };
    XMLHttpRequest.prototype.open = function (method, url, async, user, password) {
        if (async === void 0) { async = true; }
        if (this._readyState !== ReadyState.UNSENT) {
            throw new Error("Error Status ".concat(this._readyState, "!"));
        }
        this._method = method;
        this._url = url;
        this._async = async;
        this._connection = new URL(this._url).openConnection();
        if (this._connection instanceof HttpsURLConnection) {
            this._connection.setHostnameVerifier(TrustAnyHostnameVerifier);
            this._connection.setSSLSocketFactory(SSLSocketFactory);
        }
        this._connection.setRequestMethod(this._method);
        this._connection.setDoOutput(true);
        this._connection.setDoInput(true);
        this._connection.setConnectTimeout(this._connectTimeout);
        this._connection.setReadTimeout(this._readTimeout);
        this.setRequestHeader('X-Requested-With', 'XMLHttpRequest');
        this.setReadyState(ReadyState.OPENED);
    };
    XMLHttpRequest.prototype.send = function (body) {
        var _this = this;
        for (var header in this._requestHeaders) {
            this._connection.setRequestProperty(header, this._requestHeaders[header]);
        }
        if (this._readyState !== ReadyState.OPENED) {
            throw new Error("Error Status ".concat(this._readyState, "!"));
        }
        var future = executor.submit(new Callable({ call: function () { return _this._send(body); } }));
        if (!this._async) {
            future.get(this._connectTimeout + this._readTimeout + 100, TimeUnit.MILLISECONDS);
        }
        return future;
    };
    XMLHttpRequest.prototype.get = function () {
        if (this._response === undefined) {
            switch (this._responseType) {
                case "json":
                    return this._response = JSON.parse(this._responseText);
                case "text":
                    return this._response = this._responseText;
                default:
                    throw Error("Unsupport ResponseType: ".concat(this._responseType, " !"));
            }
        }
        return this._response;
    };
    XMLHttpRequest.prototype.abort = function () {
        var _a;
        this._connection.disconnect();
        (_a = this.onabort) === null || _a === void 0 ? void 0 : _a.call(this);
    };
    XMLHttpRequest.prototype._send = function (body) {
        var _a, _b, _c;
        try {
            this._connection.connect();
            (_a = this.onloadstart) === null || _a === void 0 ? void 0 : _a.call(this);
            if (body) {
                var bodyType = Object.prototype.toString.call(body);
                if (typeof body !== "string") {
                    throw new Error("body(".concat(bodyType, ") must be string!"));
                }
                var out = this._connection.getOutputStream();
                out.write(new JavaString(body).getBytes(UTF_8));
                out.flush();
                out.close();
            }
            this.setReadyState(ReadyState.LOADING);
            (_b = this.onload) === null || _b === void 0 ? void 0 : _b.call(this);
            this._status = this._connection.getResponseCode();
            this._statusText = this._connection.getResponseMessage();
            if (this._status >= 0 && this._status < 300) {
                this._responseText = this.readOutput(this._connection.getInputStream());
            }
            else if (this._status >= 300 && this._status < 400) {
                this._responseURL = this.getResponseHeader('Location');
            }
            else {
                this._responseText = this.readOutput(this._connection.getErrorStream());
            }
            this.setResponseHeaders(this._connection.getHeaderFields());
            (_c = this.onloadend) === null || _c === void 0 ? void 0 : _c.call(this);
        }
        catch (ex) {
            if (ex instanceof SocketTimeoutException && this.ontimeout) {
                return this.ontimeout(ex);
            }
            else if (this.onerror) {
                return this.onerror(ex);
            }
            throw ex;
        }
        finally {
            this._connection.disconnect();
            this.setReadyState(ReadyState.DONE);
        }
    };
    XMLHttpRequest.prototype.setResponseHeaders = function (header) {
        var _this = this;
        header.forEach(function (key, value) {
            _this._responseHeaders[key + ''] = value[value.length - 1] + '';
        });
    };
    XMLHttpRequest.prototype.setReadyState = function (state) {
        var _a;
        this._readyState = state;
        (_a = this.onreadystatechange) === null || _a === void 0 ? void 0 : _a.call(this);
    };
    XMLHttpRequest.prototype.readOutput = function (input) {
        var output = new ByteArrayOutputStream();
        var buffer = new ByteArray(1024);
        try {
            var n = void 0;
            while ((n = input.read(buffer)) != -1) {
                output.write(buffer, 0, n);
            }
            return output.toString(UTF_8);
        }
        finally {
            output.close();
        }
    };
    return XMLHttpRequest;
}());
exports.XMLHttpRequest = XMLHttpRequest;
