// @ccms/database/dist/database.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataBase = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var driver_factory_1 = require("./driver/driver-factory");
var Thread = Java.type('java.lang.Thread');
var JavaString = Java.type('java.lang.String');
var Properties = Java.type('java.util.Properties');
/**
 * 数据库封装类
 */
var DataBase = /** @class */ (function (_super) {
    tslib_1.__extends(DataBase, _super);
    function DataBase(dbConfig) {
        var _this = _super.call(this) || this;
        if (!dbConfig.url) {
            throw new Error('DataBase url can\'t be null!');
        }
        _this.dbConfig = tslib_1.__assign({ debug: global.debug }, dbConfig);
        _this.createQueryBuilder(dbConfig);
        _this.createDataSource(dbConfig);
        return _this;
    }
    DataBase.prototype.getQueryBuilder = function () {
        return this.queryBuilder;
    };
    DataBase.prototype.getDataSource = function () {
        return this.dataSource;
    };
    /**
     * 执行SQL查询
     * @param sql SQL语句
     * @param args 参数
     */
    DataBase.prototype.query = function (sql) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var startTime = Date.now();
        var result = Java.from(this.jdbcTemplate.queryForList(sql, args));
        this.debugSQL('query', sql, args, startTime);
        return result;
    };
    /**
     * 执行SQL更新
     * @param sql SQL语句
     * @param args 参数
     */
    DataBase.prototype.update = function (sql) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var startTime = Date.now();
        var result = this.jdbcTemplate.update(sql, args);
        this.debugSQL('update', sql, args, startTime);
        return result;
    };
    /**
     * 执行SQL语句
     * @param sql SQL语句
     */
    DataBase.prototype.execute = function (sql) {
        var startTime = Date.now();
        this.jdbcTemplate.execute(sql);
        this.debugSQL('execute', sql, [], startTime);
    };
    DataBase.prototype.close = function () {
        //@ts-ignore
        this.dataSource.close();
    };
    DataBase.prototype.createQueryBuilder = function (dbConfig) {
        this.queryBuilder = driver_factory_1.DriverFactory.create(dbConfig);
    };
    DataBase.prototype.createDataSource = function (dbConfig) {
        if (typeof dbConfig.url === "string") {
            var originClassLoader = Thread.currentThread().getContextClassLoader();
            Thread.currentThread().setContextClassLoader(base.getInstance().class.classLoader);
            var config = new this.HikariConfig();
            if (dbConfig.driverClassName) {
                config.setDriverClassName(dbConfig.driverClassName);
            }
            else {
                switch (dbConfig.type) {
                    case "h2":
                        config.setDriverClassName("org.h2.Driver");
                        break;
                    case "sqlite":
                        config.setDriverClassName("org.sqlite.JDBC");
                        break;
                    default:
                }
            }
            if (dbConfig.username) {
                config.setUsername(dbConfig.username);
            }
            if (dbConfig.password) {
                config.setPassword(dbConfig.password);
            }
            config.setJdbcUrl(dbConfig.url);
            if (dbConfig.properties) {
                var properties = new Properties();
                for (var key in dbConfig.properties) {
                    properties.setProperty(key, dbConfig.properties[key]);
                }
                config.setDataSourceProperties(properties);
            }
            this.debug('createDataSource from config ' + JSON.stringify(dbConfig));
            this.dataSource = new this.HikariDataSource(config);
            Thread.currentThread().setContextClassLoader(originClassLoader);
        }
        else {
            this.dataSource = dbConfig.url;
        }
        this.jdbcTemplate = new this.JdbcTemplate(this.dataSource);
    };
    DataBase.prototype.debug = function (message) {
        if (this.dbConfig.debug) {
            console.log(message);
        }
    };
    DataBase.prototype.debugSQL = function (action, sql, args, startTime) {
        this.debug(JavaString.format("\n[DB] ".concat(action, " \nSQL  : ").concat(sql.replace(/\?/ig, '%s'), " \nCOST : ").concat(Date.now() - startTime, "ms"), args));
    };
    tslib_1.__decorate([
        (0, container_1.JSClass)('com.zaxxer.hikari.HikariDataSource'),
        tslib_1.__metadata("design:type", Object)
    ], DataBase.prototype, "HikariDataSource", void 0);
    tslib_1.__decorate([
        (0, container_1.JSClass)('com.zaxxer.hikari.HikariConfig'),
        tslib_1.__metadata("design:type", Object)
    ], DataBase.prototype, "HikariConfig", void 0);
    tslib_1.__decorate([
        (0, container_1.JSClass)('org.springframework.jdbc.core.JdbcTemplate'),
        tslib_1.__metadata("design:type", Object)
    ], DataBase.prototype, "JdbcTemplate", void 0);
    DataBase = tslib_1.__decorate([
        (0, container_1.injectable)(),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], DataBase);
    return DataBase;
}(api_1.database.DataBase));
exports.DataBase = DataBase;
