// @ccms/database/dist/manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataBaseManager = void 0;
var tslib_1 = require("tslib");
var api_1 = require("@ccms/api");
var container_1 = require("@ccms/container");
var database_1 = require("./database");
var DataBaseManager = /** @class */ (function (_super) {
    tslib_1.__extends(DataBaseManager, _super);
    function DataBaseManager() {
        var _this = _super.call(this) || this;
        _this.databases = new Map();
        process.on('exit', function () { return _this.shutdown(); });
        return _this;
    }
    DataBaseManager.prototype.startWebManager = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.webManager = this.Server.createWebServer(args);
        this.webManager.start();
    };
    DataBaseManager.prototype.stopWebManager = function () {
        if (this.webManager) {
            this.webManager.stop();
            this.webManager.shutdown();
        }
    };
    DataBaseManager.prototype.shutdown = function () {
        this.stopWebManager();
        this.disable();
    };
    /**
     * 设置主数据库
     * @param mainDatabase 主数据库
     */
    DataBaseManager.prototype.setMainDatabase = function (mainDatabase) {
        this.mainDatabase = mainDatabase;
    };
    /**
     * 获得主数据库
     * Get MainDatabase
     */
    DataBaseManager.prototype.getMainDatabase = function () {
        return this.mainDatabase;
    };
    /**
     * 创建数据库
     * Create A Database Instance
     * @param name 数据库名称 用于代码 database Name use at code
     * @param config 数据库配置
     */
    DataBaseManager.prototype.createDatabase = function (name, config) {
        var _this = this;
        return Java.synchronized(function () {
            if (!_this.databases.has(name)) {
                _this.databases.set(name, new database_1.DataBase(config));
            }
            return _this.databases.get(name);
        }, this.databases)();
    };
    DataBaseManager.prototype.removeDatabase = function (name) {
        if (this.databases.has(name)) {
            this.databases.get(name).close();
        }
        return this.databases.delete(name);
    };
    DataBaseManager.prototype.getDatabase = function (name) {
        return this.databases.get(name);
    };
    DataBaseManager.prototype.disable = function () {
        this.databases.forEach(function (db) { return db.close(); });
        this.databases.clear();
    };
    tslib_1.__decorate([
        (0, container_1.JSClass)('org.h2.tools.Server'),
        tslib_1.__metadata("design:type", Object)
    ], DataBaseManager.prototype, "Server", void 0);
    DataBaseManager = tslib_1.__decorate([
        (0, container_1.provideSingleton)(api_1.database.DataBaseManager),
        tslib_1.__metadata("design:paramtypes", [])
    ], DataBaseManager);
    return DataBaseManager;
}(api_1.database.DataBaseManager));
exports.DataBaseManager = DataBaseManager;
