// @ccms/bukkit/dist/internal/chat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
/*global Java, base, module, exports, require*/
var bukkitChatInvoke;
/**
 * 获取NMS类
 */
var BukkitChatInvoke = /** @class */ (function () {
    function BukkitChatInvoke(nmsVersion) {
        this.nmsVersion = nmsVersion;
        this.downgrade = false;
    }
    BukkitChatInvoke.prototype.init = function () {
        var e_1, _a;
        try {
            try {
                this.ComponentSerializer = Java.type('net.md_5.bungee.chat.ComponentSerializer');
                this.RemapUtils = Java.type('catserver.server.remapper.RemapUtils');
            }
            catch (ex) {
            }
            var nmsChatSerializerClass = this.getNmsChatSerializerClass();
            this.nmsChatSerializerMethodName = this.getNmsChatSerializerMethodName(nmsChatSerializerClass);
            this.ChatSerializer = Java.type(nmsChatSerializerClass.getName());
            var packetTypeClass = this.getPacketPlayOutChatClass();
            this.PacketPlayOutChat = Java.type(packetTypeClass.getName());
            var constructors = packetTypeClass.constructors;
            try {
                for (var _b = tslib_1.__values(Java.from(constructors)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var constructor = _c.value;
                    var parameterTypes = constructor.parameterTypes;
                    if (parameterTypes.length === 2 || parameterTypes.length === 3) {
                        var nmsChatMessageTypeClass = parameterTypes[1];
                        if (nmsChatMessageTypeClass.isEnum()) {
                            this.chatMessageTypes = nmsChatMessageTypeClass.getEnumConstants();
                            break;
                        }
                        else if (nmsChatMessageTypeClass.getName() == 'int') {
                            break;
                        }
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            var playerConnectionField = this.getPlayerConnectionField();
            this.playerConnectionFieldName = playerConnectionField.getName();
            this.sendPacketMethodName = this.getSendPacketMethodName(playerConnectionField.getType());
        }
        catch (ex) {
            org.bukkit.Bukkit.getConsoleSender().sendMessage("\u00A76[\u00A7cMS\u00A76][\u00A7bbukkit\u00A76][\u00A7achat\u00A76] \u00A7cNMS Inject Error \u00A74".concat(ex, " \u00A7cDowngrade to Command Mode..."));
            this.downgrade = true;
        }
    };
    BukkitChatInvoke.prototype.nmsCls = function (name) {
        return base.getClass(['net.minecraft.server', this.nmsVersion, name].join('.'));
    };
    BukkitChatInvoke.prototype.remapMethod = function (clazz, origin, test, params) {
        try {
            return clazz.getMethod(origin, params);
        }
        catch (ex) {
            if (this.RemapUtils) {
                return clazz.getMethod(this.RemapUtils.mapMethod(clazz, origin, params), params);
            }
            else {
                return clazz.getMethod(test, params);
            }
        }
    };
    BukkitChatInvoke.prototype.remapFieldName = function (clazz, origin, test) {
        try {
            return clazz.getField(origin);
        }
        catch (ex) {
            if (this.RemapUtils) {
                return clazz.getField(this.RemapUtils.mapFieldName(clazz, origin));
            }
            else {
                return clazz.getField(test);
            }
        }
    };
    BukkitChatInvoke.prototype.json = function (sender, json) {
        if (this.downgrade) {
            return sender.spigot().sendMessage(this.ComponentSerializer.parse(json));
        }
        else {
            this.send(sender, json, 0);
            return false;
        }
    };
    BukkitChatInvoke.prototype.send = function (sender, json, type) {
        this.sendPacket(sender, this.getPacketPlayOutChat(sender, json, type));
    };
    BukkitChatInvoke.prototype.sendPacket = function (player, p) {
        player.handle[this.playerConnectionFieldName][this.sendPacketMethodName](p);
    };
    return BukkitChatInvoke;
}());
var BukkitChatInvokeBase = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitChatInvokeBase, _super);
    function BukkitChatInvokeBase() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BukkitChatInvokeBase.prototype.getSendPacketMethodName = function (playerConnectionClass) {
        return this.remapMethod(playerConnectionClass, 'sendPacket', 'func_179290_a', this.getPacketClass()).getName();
    };
    BukkitChatInvokeBase.prototype.getNmsChatSerializerMethodName = function (nmsChatSerializerClass) {
        var nmsChatSerializerMethod = this.remapMethod(nmsChatSerializerClass, 'a', 'func_150699_a', base.getClass('java.lang.String'));
        return nmsChatSerializerMethod.getName();
    };
    BukkitChatInvokeBase.prototype.getPacketPlayOutChat = function (sender, json, type) {
        return new this.PacketPlayOutChat(this.ChatSerializer[this.nmsChatSerializerMethodName](json), type);
    };
    BukkitChatInvokeBase.prototype.getNmsChatSerializerClass = function () {
        return this.nmsCls("ChatSerializer");
    };
    BukkitChatInvokeBase.prototype.getPacketPlayOutChatClass = function () {
        return this.nmsCls("PacketPlayOutChat");
    };
    BukkitChatInvokeBase.prototype.getPlayerConnectionField = function () {
        return this.remapFieldName(this.nmsCls('EntityPlayer'), 'playerConnection', 'field_71135_a');
    };
    BukkitChatInvokeBase.prototype.getPacketClass = function () {
        return this.nmsCls('Packet');
    };
    return BukkitChatInvokeBase;
}(BukkitChatInvoke));
var BukkitChatInvoke_1_7_10 = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitChatInvoke_1_7_10, _super);
    function BukkitChatInvoke_1_7_10() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return BukkitChatInvoke_1_7_10;
}(BukkitChatInvokeBase));
var BukkitChatInvoke_1_8 = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitChatInvoke_1_8, _super);
    function BukkitChatInvoke_1_8() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BukkitChatInvoke_1_8.prototype.getPacketPlayOutChat = function (sender, json, type) {
        return new this.PacketPlayOutChat(this.ChatSerializer[this.nmsChatSerializerMethodName](json), this.chatMessageTypes[type]);
    };
    BukkitChatInvoke_1_8.prototype.getNmsChatSerializerClass = function () {
        return this.nmsCls("IChatBaseComponent$ChatSerializer");
    };
    return BukkitChatInvoke_1_8;
}(BukkitChatInvoke_1_7_10));
var BukkitChatInvoke_1_16_5 = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitChatInvoke_1_16_5, _super);
    function BukkitChatInvoke_1_16_5() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BukkitChatInvoke_1_16_5.prototype.getPacketPlayOutChat = function (sender, json, type) {
        return new this.PacketPlayOutChat(this.ChatSerializer[this.nmsChatSerializerMethodName](json), this.chatMessageTypes[type], sender.getUniqueId());
    };
    return BukkitChatInvoke_1_16_5;
}(BukkitChatInvoke_1_8));
var BukkitChatInvoke_1_17_1 = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitChatInvoke_1_17_1, _super);
    function BukkitChatInvoke_1_17_1() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BukkitChatInvoke_1_17_1.prototype.getPacketPlayOutChatClass = function () {
        return base.getClass('net.minecraft.network.protocol.game.PacketPlayOutChat');
    };
    BukkitChatInvoke_1_17_1.prototype.getNmsChatSerializerClass = function () {
        return base.getClass('net.minecraft.network.chat.IChatBaseComponent$ChatSerializer');
    };
    BukkitChatInvoke_1_17_1.prototype.getPlayerConnectionField = function () {
        return base.getClass('net.minecraft.server.level.EntityPlayer').getField('b');
    };
    BukkitChatInvoke_1_17_1.prototype.getPacketClass = function () {
        return base.getClass('net.minecraft.network.protocol.Packet');
    };
    return BukkitChatInvoke_1_17_1;
}(BukkitChatInvoke_1_16_5));
var BukkitChatInvoke_1_18_2 = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitChatInvoke_1_18_2, _super);
    function BukkitChatInvoke_1_18_2() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BukkitChatInvoke_1_18_2.prototype.getSendPacketMethodName = function (playerConnectionClass) {
        return playerConnectionClass.getMethod('a', this.getPacketClass()).getName();
    };
    return BukkitChatInvoke_1_18_2;
}(BukkitChatInvoke_1_17_1));
var BukkitChatInvoke_1_19 = /** @class */ (function (_super) {
    tslib_1.__extends(BukkitChatInvoke_1_19, _super);
    function BukkitChatInvoke_1_19() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    BukkitChatInvoke_1_19.prototype.getPacketPlayOutChatClass = function () {
        return base.getClass('net.minecraft.network.protocol.game.ClientboundSystemChatPacket');
    };
    BukkitChatInvoke_1_19.prototype.getPacketPlayOutChat = function (sender, json, type) {
        return new this.PacketPlayOutChat(this.ChatSerializer[this.nmsChatSerializerMethodName](json), type);
    };
    return BukkitChatInvoke_1_19;
}(BukkitChatInvoke_1_18_2));
try {
    var Bukkit = Java.type('org.bukkit.Bukkit');
    // @ts-ignore
    var nmsVersion = Bukkit.getServer().class.name.split('.')[3];
    var nmsSubVersion = nmsVersion.split("_")[1];
    if (nmsSubVersion >= 19) {
        bukkitChatInvoke = new BukkitChatInvoke_1_19(nmsVersion);
    }
    else if (nmsSubVersion >= 18) {
        bukkitChatInvoke = new BukkitChatInvoke_1_18_2(nmsVersion);
    }
    else if (nmsSubVersion >= 17) {
        bukkitChatInvoke = new BukkitChatInvoke_1_17_1(nmsVersion);
    }
    else if (nmsSubVersion >= 16) {
        bukkitChatInvoke = new BukkitChatInvoke_1_16_5(nmsVersion);
    }
    else if (nmsSubVersion >= 8) {
        bukkitChatInvoke = new BukkitChatInvoke_1_8(nmsVersion);
    }
    else {
        bukkitChatInvoke = new BukkitChatInvoke_1_7_10(nmsVersion);
    }
    bukkitChatInvoke.init();
}
catch (ex) {
}
var chat = {
    invoke: bukkitChatInvoke,
    json: bukkitChatInvoke.json.bind(bukkitChatInvoke),
    send: bukkitChatInvoke.send.bind(bukkitChatInvoke)
};
exports.default = chat;
